---    
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---
        
```{r setup0, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', fig.height = 9, fig.width = 12, bootstrap.show.code=FALSE)
```

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Weiner RNA-seq - DE analysis"
clientname="Dan Hu"
clientemail=""
labPI="Weiner"
analystname="Mary Piper"
analystemail="piper@hsph.harvard.edu"
```
---
        
# Weiner RNA-Seq Differential Expression Report
        
RNA-Seq differential expression report for RNA-Seq on four types of CD4+ T cells (DP, Th17, Th1, DN) that were isolated from age, sex, and race-matched healthy human donors and patients. The CD4+ T cells were separated into the four cell types by levels of IL-17 and IFN-g using FACS. 

Client: `r clientname`, `r labPI` group.  

Analysts: `r analystname` (`r analystemail`)

The most recent update of this html document occurred: `r date()`

---
        
# Overview
        
```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(RColorBrewer)
library(knitr)
library(tidyr)
library(reshape)
library(rmarkdown)
library(dplyr)
library(ggdendro)
library(grid)
library(reshape)
library(gridExtra)
library(Biobase)
library(scales)
library(genefilter)
library(biomaRt)
source('~/Dropbox/HBC consults/2015_11_09_weiner/revigo.R')

#Bringing in run 1 (batch 1)
sample_names1 <- c("S1-1", "S2-1", "S3-1", "S4-1", "S5-1", "S6-1", "S7-1", "S8-1", "S9-1", "S10-1", "S11-1", "S12-1", "S13-1", "S14-1", "S15-1", "S16-1", "S17-1", "S18-1", "S19-1", "S20-1", "S21-1", "S22-1", "S23-1", "S24-1")
project_summary1 = "~/Dropbox/HBC consults/2015_11_09_weiner/run1/project-summary_nobatch2.csv"
counts_file1 = "~/Dropbox/HBC consults/2015_11_09_weiner/run1/data/combined.counts_nobatch2.txt"
path_results1 = "~/Dropbox/HBC consults/2015_11_09_weiner/run1/report"
ann.genes1 <- read.table(file.path('~/Dropbox/HBC consults/2015_11_09_weiner/run1/data/annotated_combined.counts_nobatch2'), header=T, sep="\t", row.names=1, as.is=T)
colnames(ann.genes1) <- c(sample_names1,"symbol")
#ann.genes1 <- ann.genes1[,c(1:24,37)]
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
               "#0072B2", "#D55E00", "#CC79A7")
summarydata1 = data.frame(read.table(project_summary1, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata1$batch <- rep(1,24)
summarydata1$celltype <- c("DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","Th17", "DP", "Th1", "DN", "DP", "Th17", "Th1", "DN", "DP", "Th17", "Th1", "DN")
summarydata1$Name = rownames(summarydata1)
summarydata1 = summarydata1[order(summarydata1$Name),]
summarydata1 <- summarydata1[,c(1:25,28,31:33)]
rownames(summarydata1) <- sample_names1
counts1 = read.table(counts_file1, header=TRUE, row.names="id", check.names=FALSE)
counts1 = counts1[, order(colnames(counts1))]
colnames(counts1) <- sample_names1
total_counts1 <- colSums (counts1, na.rm = FALSE, dims = 1)
y = DGEList(counts=counts1)
y = calcNormFactors(y)
normalized_counts1 = cpm(y, normalized.lib.sizes=TRUE)
colnames(counts1) = gsub(".counts", "", colnames(counts1))
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
                  "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
                  "rRNA_rate", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
                  "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
                  "complexity", "X5.3.bias")
meta1 = summarydata1[, !colnames(summarydata1) %in% known_columns, drop=FALSE]
metadata1 = subset(meta1, select=samplegroup)
metadata1$celltype <- c("DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","Th17", "DP", "Th1", "DN", "DP", "Th17", "Th1", "DN", "DP", "Th17", "Th1", "DN")
metadata1$disease <- c("patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy", "patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy", "patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy")
metadata1$batch <- rep(1, 24)

qc1 <- read.table(file.path(path_results1, "metrics", "metrics_nobatch2.tsv"),
                  header=T, sep="\t", check.names=F,
                  colClasses=list("sample"="character"))
rownames(qc1) = sample_names1
qc1$Mapped_reads_pct = as.numeric(gsub("%", "", qc1$Mapped_reads_pct))
qc1$Duplicates_pct = as.numeric(gsub("%", "", qc1$Duplicates_pct))

#Bringing in run 2
sample_names2 <- c("S1-2", "S2-2", "S3-2", "S4-2", "S5-2", "S6-2", "S7-2", "S8-2", "S9-2", "S10-2", "S11-2", "S12-2", "S13-2", "S14-2", "S15-2", "S16-2", "S17-2", "S18-2", "S19-2", "S20-2", "S21-2", "S22-2", "S23-2", "S24-2")
project_summary2 = "~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq2/project-summary2.csv"
counts_file2 = "~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq2/combined.counts2"
path_results2 = "~/Dropbox/HBC consults/2015-11-09_weiner/run2/2015-12-13_weiner-rnaseq2/report"
ann.genes2 <- read.table(file.path("~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq2/annotated_combined.counts2"), header=T, sep="\t", row.names=1, as.is=T)
colnames(ann.genes2) <- c(sample_names2, "symbol")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
               "#0072B2", "#D55E00", "#CC79A7")
summarydata2 = data.frame(read.table(project_summary2, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata2$Name = rownames(summarydata2)
summarydata2 = summarydata2[order(summarydata2$Name),]
rownames(summarydata2) <- sample_names2
counts2 = read.table(counts_file2, header=TRUE, row.names="id", check.names=FALSE)
counts2 = counts2[, order(colnames(counts2))]
colnames(counts2) <- sample_names2
total_counts2 <- colSums (counts2, na.rm = FALSE, dims = 1)
y = DGEList(counts=counts2)
y = calcNormFactors(y)
normalized_counts2 = cpm(y, normalized.lib.sizes=TRUE)
colnames(counts2) = gsub(".counts", "", colnames(counts2))
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
                  "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
                  "rRNA_rate", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
                  "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
                  "complexity", "X5.3.bias")
meta2 = summarydata2[, !colnames(summarydata2) %in% known_columns, drop=FALSE]
metadata2 = subset(meta2, select=samplegroup)
metadata2$celltype <- c("DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","Th17", "DP", "Th1", "DN", "DP", "Th17", "Th1", "DN", "DP", "Th17", "Th1", "DN")
metadata2$disease <- c("patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy", "patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy", "patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy")
metadata2$batch <- rep(2, 24)

qc2 <- read.table("~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq2/report/metrics/metrics.tsv", header=T, sep="\t", check.names=F)
rownames(qc2) = sample_names2
qc2$Mapped_reads_pct = as.numeric(gsub("%", "", qc2$Mapped_reads_pct))
qc2$Duplicates_pct = as.numeric(gsub("%", "", qc2$Duplicates_pct))

#Bringing in run 3
sample_names3 <- c("S1-3", "S2-3", "S3-3", "S4-3", "S5-3", "S6-3", "S7-3", "S8-3", "S9-3", "S10-3", "S11-3", "S12-3", "S13-3", "S14-3", "S15-3", "S16-3", "S17-3", "S18-3", "S19-3", "S20-3", "S21-3", "S22-3", "S23-3", "S24-3")
project_summary3 = "~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq3/project-summary3.csv"
counts_file3 = "~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq3/combined.counts3"
path_results3 = "~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq3/report/"
ann.genes3 <- read.table(file.path("~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq3/annotated_combined.counts3"), header=T, sep="\t", row.names=1, as.is=T)
colnames(ann.genes3) = c(sample_names3, "symbol")
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
               "#0072B2", "#D55E00", "#CC79A7")
summarydata3 = data.frame(read.table(project_summary3, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata3$Name = rownames(summarydata3)
summarydata3 = summarydata3[order(summarydata3$Name),]
rownames(summarydata3) <- sample_names3
counts3 = read.table(counts_file3, header=TRUE, row.names="id", check.names=FALSE)
counts3 = counts3[, order(colnames(counts3))]
colnames(counts3) <- sample_names3
total_counts3 <- colSums (counts3, na.rm = FALSE, dims = 1)
y = DGEList(counts=counts3)
y = calcNormFactors(y)
normalized_counts3 = cpm(y, normalized.lib.sizes=TRUE)
colnames(counts3) = gsub(".counts", "", colnames(counts3))
# this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
                  "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
                  "rRNA_rate", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
                  "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
                  "complexity", "X5.3.bias")
meta3 = summarydata3[, !colnames(summarydata3) %in% known_columns, drop=FALSE]
metadata3 = subset(meta3, select=samplegroup)
metadata3$celltype <- c("DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","Th17", "DP", "Th1", "DN", "DP", "Th17", "Th1", "DN", "DP", "Th17", "Th1", "DN")
metadata3$disease <- c("patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy", "patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy", "patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy")
metadata3$batch <- rep(3, 24)
qc3 <- read.table("~/Dropbox/HBC consults/2015_11_09_weiner/run2/2015-12-13_weiner-rnaseq3/report/metrics/metrics.tsv", header=T, sep="\t", check.names=F)
rownames(qc3) = sample_names3
qc3$Mapped_reads_pct = as.numeric(gsub("%", "", qc3$Mapped_reads_pct))
qc3$Duplicates_pct = as.numeric(gsub("%", "", qc3$Duplicates_pct))


#Merging all runs for analysis
counts_merged <- cbind(counts1, counts2, counts3)
metadata_merged <- rbind(metadata1, metadata2, metadata3)
colnames(counts_merged) <- rownames(metadata_merged)
summarydata_merged <- rbind(summarydata1, summarydata2, summarydata3)
qc_merged <- rbind(qc1, qc2, qc3)

```

```{r heatmap-function}
get_heatmap_fn = function(summarydata_merged) {
        # return the pheatmap function with or without metadata   
        if(ncol(metadata_merged) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=metadata_merged, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn = get_heatmap_fn(summarydata_merged)
```

#Quality Control

## Total reads
The total reads are the number of input reads in each sample. The number of reads for each sample varies greatly (between <5 million to >30 million), and the second and third runs generated larger numbers of reads. This is expected based on the previous analysis of the first run.

```{r total-reads}
# Quality control metrics
metrics = c("sample", "Total_reads" ,"Mapped_reads_pct", "Duplicates_pct",
            "offtarget",
            "%GC", "Sequence_length", "Median_insert_size")

ggplot(qc_merged, aes(x=rownames(qc_merged), y=Total_reads/1e6)) +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        geom_bar(stat = 'identity') +
        ylab("Million reads")
```

## Mapped reads
The number of mapped reads is low across samples, due to the low number of input reads. The additional sequencing runs more than tripled the number of mapped reads per sample. However, there is still a large amount of variation in the number of mapped reads per sample, with S13 <10 million and other samples with >50 million mapped reads.

```{r mapped-plot}

ggplot(summarydata_merged, aes(x=rownames(summarydata_merged), y=Mapped/1e6)) +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        geom_bar(stat="identity") +
        ylab("Million mapped reads") + xlab("")
```

## Genomic mapping rate
The genomic mapping rate represents the percentage of reads mapping to the reference genome. Low mapping rates are indicative of sample contamination, poor sequencing quality or other artifacts.

The percent of input reads mapped is generally decent (75-90%) for nearly all of the samples.

```{r mapping-rate-plot}
ggplot(summarydata_merged, aes(x=rownames(summarydata_merged), y=Mapping.Rate)) +
        geom_bar(stat="identity") +
        ylab("mapping rate (mapped_reads/total_reads") + xlab("") +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90))
```

## Unique mapping rate
The unique mapping rate is low, with only ~20% of the reads mapping to unique locations on the genome. High rates of duplication were identified during QC. Also, the low unique mapping rate indicates the libraries are of low complexity, which is not surprising due to the low number of input reads. These results are similar to what we found during the previous analysis.

```{r unique-rate-plot}
dd = data.frame(Name=names(counts_merged), Unique=colSums(counts_merged), Mapped=summarydata_merged[,"Mapped"])
ggplot(dd, aes(x=Name, y=Unique/Mapped)) +
        geom_bar(stat="identity") +
        ylab("unique mapping rate") + xlab("") +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90))
```

## Number of genes detected
The addition of sequencing runs allowed for the detection of a greater number of genes, but three samples, S13, S21, and S22 show low numbers of genes detected. Low genes detected for S13 is likely due to the low number of reads, but S21 and S22 are likely due to low complexity of the libraries since both samples had ~50 million mapped reads. 

```{r genes-detected-plot}
dd = data.frame(Name=names(counts_merged), Genes.Detected = colSums(counts_merged > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
        geom_bar(stat="identity") +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        ylab("genes detected") + xlab("")
```

## Gene detection saturation
Based on the low complexity of the libraries, gene detection appears saturated at 10 million mapped reads. Also, samples vary substantially in number of genes detected based on the number of reads mapped, indicating large variation in the complexity of the libraries. Variation in library complexity may complicate the differential expression analysis. 

```{r saturation-plot}
dd = data.frame(Mapped=summarydata_merged$Mapped, Genes.Detected = colSums(counts_merged > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
geom_point() +
theme_bw(base_size=10) +
theme(panel.grid.major = element_line(size = .5, color = "grey"),
axis.text.x = element_text(angle=90)) +
ylab("genes detected") + xlab("reads mapped")
```

## Exonic mapping rate
Most samples have low exonic mapping rates of between 30-50%. This is expected for the NuGen Ovation RNA-Seq kits.
```{r exonic-mapping-plot}
ggplot(summarydata_merged, aes(x=rownames(summarydata_merged), y=Exonic.Rate)) +
geom_bar(stat="identity") +
theme_bw(base_size=10) +
theme(panel.grid.major = element_line(size = .5, color = "grey"),
axis.text.x = element_text(angle=90)) +
ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
The rRNA mapping rate is very high, generally between 30-40% of mapped reads, similar to the previous analysis. Sample S22 has the highest percentage of rRNA, which could be part of the reason for the low complexity of that library.

```{r rRNA-rate-plot}
ggplot(summarydata_merged, aes(x=rownames(summarydata_merged), y=rRNA_rate)) +
geom_bar(stat="identity") +
theme_bw(base_size=10) +
theme(panel.grid.major = element_line(size = .5, color = "grey"),
axis.text.x = element_text(angle=90)) +
ylab("rRNA rate") + xlab("")
```

## Estimated fragment length of paired-end reads
Similar fragment lengths of 150 base pairs were estimated across samples. S13 spikes to 200 bp, which could be due to the size of the mtRNA, if present in high proportions in the sample.
```{r fragment-length-plot}
ggplot(summarydata_merged, aes(x=rownames(summarydata_merged), y=Fragment.Length.Mean)) +
geom_bar(stat="identity") +
theme_bw(base_size=10) +
theme(panel.grid.major = element_line(size = .5, color = "grey"),
axis.text.x = element_text(angle=90)) +
ylab("fragment length") + xlab("")
```

## Boxplot of log10 counts per gene
This plot shows a lot of variation between samples for the general spread of the data.
```{r boxplot-raw}
melted = melt(counts_merged)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
theme_bw(base_size=10) +
theme(panel.grid.major = element_line(size = .5, color = "grey"),
axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
The spread of the log10 TMM-normalized counts per gene data should be similar for every sample. Optimal data would display similar boxplots for each sample without many outliers. 

This plot shows quite a bit of variation between samples for the general spread of the data. S13 sample is a bit of an outlier, in addition to S21 and S22.

Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts_merged)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
theme_bw(base_size=10) +
theme(panel.grid.major = element_line(size = .5, color = "grey"),
axis.text.x = element_text(angle=90)) + xlab("")

```

## Density of log10 TMM-normalized counts
Density of normalized counts per gene should be very similar between samples, with no large shifts in any of the sample curves.

Our data shows a great degree of variation between samples for all genes, which may skew the differential expression results. Generally, these curves are similar, with a single large outlier (S13). S13 is likely an outlier due to the small number of reads; therefore, for this sample it is difficult to detect many of the lowly expressed genes.

```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
geom_density() +
theme_bw(base_size=10) +
theme(panel.grid.major = element_line(size = .5, color = "grey"),
axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Pearson) heatmap of TMM-normalized counts
Correlation of the same sample between different sequencing runs is very good, but the correlation between samples of the same group is poor. There is very little clustering within cell types groups.
```{r pearson-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="pearson"))
```

## Correlation (Spearman) heatmap of TMM-normalized counts
The Spearman correlation heatmap is more resistant to outliers than the Pearson heatmap (above), but again, the cell type groups do not cluster well.
```{r spearman-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

## PCA plot of TMM-normalized counts
Principal components analysis is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expression levels for genes and transforms it in principal component space, reducing each sample into one point. Thereby, we can separate samples by expression variation, and identify potential sample outliers. The PCA plot is another way to look at how samples are clustering. 

In this PCA plot the same sample across sequencing runs cluster well, but the samples of the same group do not cluster well.
```{r pca}
metadata_merged$name <- row.names(metadata_merged)

pca_matrix <- prcomp(t(normalized_counts))$x
df <- cbind(metadata_merged, pca_matrix[, c("PC1", "PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
0.3)) + ggtitle("PC1 vs PC2 :: normalized counts")
```


## Heatmap of top 30 most expressed genes
Many of the top expressers are mitochondrial genes, as identified in the previous analysis. 
```{r top-count-genes, results='asis'}
metadata_merged <- metadata_merged[,1:2]
select = order(rowMeans(counts_merged),decreasing=TRUE)[1:30]
heatmap_fn(counts_merged[select,])
```

## Heatmap by concordance correlation coefficient
Similarly, no clustering by sample group.

http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

```{r propcor-heatmap}
propcor = function(x, y) {
x = log(x + 0.1)
y = log(y + 0.1)
num = 2 * cov(x, y)
denom = var(x) + var(y)
return(num/denom)}

do_propcor = function(x) {
mat = list()
for(i in seq_len(ncol(x))) {
for(j in seq_len(ncol(x))) {
x2 = x[, i]
y2 = x[, j]
mat = c(mat, propcor(x2, y2)) } }
mat = unlist(mat)
mat = matrix(mat, ncol(x), ncol(x))
colnames(mat) = colnames(x)
rownames(mat) = colnames(x)
return(mat)}

heatmap_fn(do_propcor(normalized_counts))
```

#Analysis without rRNA and mitochondrial (mt) genes 
Due to the high levels of reads aligning to rRNA and mt genes, clustering analysis was performed after removal of count data for these genes. In the entire dataset, reads aligning to the mitochondrial and rRNA genes account for over 62 million of the reads aligning to exons in the dataset, leaving 100 million total reads aligning to protein-coding genes (700,000 - 6,700,000 reads per combined sample). 
```{r no rrna_marker_counts}
human = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
dataset="hsapiens_gene_ensembl",
host = "jul2015.archive.ensembl.org")
conversions = getBM(attributes=c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
mart=human)
rrna_biotypes = c("rRNA", "Mt_rRNA", "misc_RNA", "snRNA", "snoRNA",
"tRNA", "Mt_tRNA")

rrna_genes <- unique(subset(conversions, gene_biotype %in% rrna_biotypes)$ensembl_gene_id)

mt_genes <- c("ENSG00000210049","ENSG00000211459","ENSG00000210077","ENSG00000210082","ENSG00000209082","ENSG00000198888","ENSG00000210100","ENSG00000210107","ENSG00000210112","ENSG00000198763","ENSG00000210117","ENSG00000210127","ENSG00000210135","ENSG00000210140","ENSG00000210144","ENSG00000198804","ENSG00000210151","ENSG00000210154","ENSG00000198712","ENSG00000210156","ENSG00000228253","ENSG00000198899","ENSG00000198938","ENSG00000210164","ENSG00000198840","ENSG00000210174","ENSG00000212907","ENSG00000198886","ENSG00000210176","ENSG00000210184","ENSG00000210191","ENSG00000198786","ENSG00000198695","ENSG00000210194","ENSG00000198727","ENSG00000210195","ENSG00000210196")

#Protein coding genes
human = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
dataset="hsapiens_gene_ensembl",
host = "jul2015.archive.ensembl.org")
protein_coding <- getBM(attributes=c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),filters = "biotype", values="protein_coding",mart=human)
pc_genes <- protein_coding[,1]

counts_pc_genes <- counts_merged[rownames(counts_merged) %in% pc_genes,]
sum_pc_genes <- colSums(counts_pc_genes, na.rm = FALSE, dims = 1)
#sum(sum_pc_genes)

#Reads aligning to mt genes
counts_mt_genes <- counts_merged[rownames(counts_merged) %in% mt_genes,]
sum_mt_genes <- colSums(counts_mt_genes, na.rm = FALSE, dims = 1)
#sum(sum_mt_genes)
#Reads aligning to rRNA genes
counts_rrna_genes <- counts_merged[rownames(counts_merged) %in% rrna_genes,]
sum_rrna_genes <- colSums(counts_rrna_genes, na.rm = FALSE, dims = 1)
#sum(sum_rrna_genes)

#Counts rrna genes not in mt genes
counts_gen_rrna <- counts_rrna_genes[!rownames(counts_rrna_genes) %in% mt_genes,]
sum_gen_rrna_genes <- colSums(counts_gen_rrna, na.rm = FALSE, dims = 1)
#Counts rrna genes in mt genes
counts_mt_rrna <- counts_rrna_genes[rownames(counts_rrna_genes) %in% mt_genes,]
sum_mt_rrna_genes <- colSums(counts_mt_rrna, na.rm = FALSE, dims = 1)

#Analysis without mt and rRNA genes
norm_counts_no_rrna <- normalized_counts[!rownames(normalized_counts) %in% rrna_genes & !rownames(normalized_counts) %in% mt_genes,]

metadata_merged$name <- row.names(metadata)
metadata_merged$celltype <- c("DP", "Th17", "Th1", "DN","DP", "Th17", "Th1", "DN","DP", "Th17", "Th1","DN","Th17", "DP", "Th1", "DN", "DP", "Th17", "Th1", "DN", "DP", "Th17", "Th1", "DN")
metadata_merged$disease <- c("patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy", "patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy", "patient", "patient", "patient", "patient", "healthy", "healthy", "healthy", "healthy")
metadata_merged$pair <- c(1,1,1,1,1,1,1,1,2,2,2,2,2,2,2,2,3,3,3,3,3,3,3,3)
```

## Heatmap of marker genes
Positive identification of the expression differences between the cell types of marker genes. This is necessary if there is any hope to identify differentially expressed genes between or within cell types.
```{r Heatmap of markers}
counts_no_rrna_merged = counts_merged[!rownames(counts_merged) %in% rrna_genes & !rownames(counts_merged) %in% mt_genes,]
y = DGEList(counts=counts_no_rrna_merged)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
ann.norm.counts <- merge(normalized_counts, ann.genes1, by=0)
row.names(ann.norm.counts) <- ann.norm.counts$Row.names
ann.norm.counts <- ann.norm.counts[,c(2:73,98)]
ann_norm_counts_no_rrna = ann.norm.counts[!rownames(ann.norm.counts) %in% rrna_genes & !rownames(ann.norm.counts) %in% mt_genes,]

marker_norm_counts <- subset(ann_norm_counts_no_rrna, symbol == "IFNG" | symbol == "IL17A")
marker_norm_counts <- marker_norm_counts[1:72]
colnames(marker_norm_counts) <- rownames(summarydata_merged)
heatmap_fn(cor(marker_norm_counts, method="spearman"))
heatmap_fn((marker_norm_counts))
```

###PCA plot - no rrna or mitochondrial genes
Still poor clustering, although better than previously. Some separation on the PC2 axis of the DN and Th17 cells from the DP and Th1 cells. 
```{r no rrna_PCA}
#mds(norm_counts_no_rrna, k=length(colnames(norm_counts_no_rrna)) - 1)

pca_matrix <- prcomp(t(norm_counts_no_rrna))$x
df <- cbind(metadata_merged, pca_matrix[, c("PC1", "PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
                                                           size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
                                                                                                                               0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")
```

## Correlation (Spearman) heatmap of TMM-normalized counts - no rrna or mitochondrial genes
The Spearman correlation heatmap is more resistant to outliers than the Pearson heatmap (above), but still, the groups do not cluster well.
```{r no rrna_heatmap}
heatmap_fn(cor(norm_counts_no_rrna, method="spearman"))
```

##PCA of the top 500 variant genes
Often the most variant genes are due to differences between sample groups. If this is the case,  you can explore this subset of genes for differential expression. However, the top 500 variant genes do not cluster well by sample group well.
```{r no rrna_top 500 variant genes-PCA}
ntop <- 500
rv <- rowVars(norm_counts_no_rrna)
select <- order(rv, decreasing=TRUE)[seq_len(ntop)]
pca <- prcomp(t(norm_counts_no_rrna[select,]))
pca_cut <- pca[["x"]]
pca_cut <- pca_cut[,1:2]
df <- cbind(metadata_merged, pca_cut[,c("PC1","PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
                                                           size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
                                                                                                                               0.3)) + ggtitle("PC1 vs PC2 :: 500 top variable genes (no rrna)")
```

##Analysis of Genes of Interest (15 genes)
Exploring the list of genes of interest. Overall, poor clustering by sample group, which makes it very unlikely that we will see significant differential expression. 

###Analysis of Genes of Interest (15 genes) - PCA
```{r genelist_setup_merged}
ann.genes <- cbind(ann.genes1[1:24], ann.genes2[1:24], ann.genes3)
y = DGEList(counts=counts_no_rrna_merged)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
ann.norm.counts <- merge(normalized_counts, ann.genes, by=0)
row.names(ann.norm.counts) <- ann.norm.counts$Row.names
ann.norm.counts <- ann.norm.counts[,c(2:73,146)]
colnames(ann.norm.counts) <- c(colnames(counts_merged), "symbol")
ann_norm_counts_no_rrna = ann.norm.counts[!rownames(ann.norm.counts) %in% rrna_genes & !rownames(ann.norm.counts) %in% mt_genes,]

ccl3 <- subset(ann_norm_counts_no_rrna, symbol == "CCL3")
ccl4 <- subset(ann_norm_counts_no_rrna, symbol == "CCL4")
ccl5 <- subset(ann_norm_counts_no_rrna, symbol == "CCL5")
ccl6 <- subset(ann_norm_counts_no_rrna, symbol == "CSF2")
ccl7 <- subset(ann_norm_counts_no_rrna, symbol == "CXCR3")
ccl8 <- subset(ann_norm_counts_no_rrna, symbol == "EPSTI1")
ccl9 <- subset(ann_norm_counts_no_rrna, symbol == "GZMB")
ccl10 <- subset(ann_norm_counts_no_rrna, symbol == "IFNG")
ccl11<- subset(ann_norm_counts_no_rrna, symbol == "IL23R")
ccl12<- subset(ann_norm_counts_no_rrna, symbol == "STAT3")
ccl13<- subset(ann_norm_counts_no_rrna, symbol == "IL3")
ccl14<- subset(ann_norm_counts_no_rrna, symbol == "RGS2")
ccl15 <- subset(ann_norm_counts_no_rrna, symbol == "STAT1")
ccl16<- subset(ann_norm_counts_no_rrna, symbol == "STAT5A")
ccl17<- subset(ann_norm_counts_no_rrna, symbol == "TBX21")
genes <- data.frame()
genes <- rbind(ccl3, ccl4, ccl5, ccl6, ccl7, ccl8, ccl9, ccl10, ccl11, ccl12, ccl13, ccl14, ccl15, ccl16, ccl17)
row.names(genes) <- genes$symbol
genes <- genes[,1:72]

#PCA plot of 15 genes
pca_matrix <- prcomp(t(genes))$x
df <- cbind(metadata_merged, pca_matrix[, c("PC1", "PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
                                                           size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3,0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")   
```

###Analysis of Genes of Interest (15 genes) - Heatmap
Poor clustering of the genes on the heatmap too. Again, the lack of clustering suggests that these genes will not be identified by differential expression expression analysis, since there is a large amount variation within sample groups.
```{r top15_heatmap}
heatmap_fn(cor(genes, method="spearman"))
heatmap_fn(genes, method="spearman", scale="row")
write.table(genes, file ="normalized_counts_15_genes_merged.txt", quote = FALSE, sep = "\t", row.names=T)

#Raw counts table
ann.genes <- cbind(ann.genes1, ann.genes2[1:24], ann.genes3[1:24])
ccl3 <- subset(ann.genes, symbol == "CCL3")
ccl4 <- subset(ann.genes, symbol == "CCL4")
ccl5 <- subset(ann.genes, symbol == "CCL5")
ccl6 <- subset(ann.genes, symbol == "CSF2")
ccl7 <- subset(ann.genes, symbol == "CXCR3")
ccl8 <- subset(ann.genes, symbol == "EPSTI1")
ccl9 <- subset(ann.genes, symbol == "GZMB")
ccl10 <- subset(ann.genes, symbol == "IFNG")
ccl11<- subset(ann.genes, symbol == "IL23R")
ccl12<- subset(ann.genes, symbol == "STAT3")
ccl13<- subset(ann.genes, symbol == "IL3")
ccl14<- subset(ann.genes, symbol == "RGS2")
ccl15 <- subset(ann.genes, symbol == "STAT1")
ccl16<- subset(ann.genes, symbol == "STAT5A")
ccl17<- subset(ann.genes, symbol == "TBX21")
counts_15_genes <- rbind(ccl3, ccl4, ccl5, ccl6, ccl7, ccl8, ccl9, ccl10, ccl11, ccl12, ccl13, ccl14, ccl15, ccl16, ccl17)
row.names(counts_15_genes) <- counts_15_genes$symbol
counts_15_genes <- counts_15_genes[, c(1:24, 26:73)]
#write.table(counts_15_genes, file ="raw_counts_15_genes.txt", quote = FALSE, sep = "\t", row.names=F)
```

# Combined results from all runs without rrna and mt genes
Since the individual samples cluster well across all runs, the sample counts will be combined for differential expression analysis. First, we ensure the count spread still looks okay between samples.
```{r setup_combined}
combined_all_counts <- counts1 + counts2 + counts3
combined_ann_genes <- ann.genes1[1:24] + ann.genes2[1:24] + ann.genes3[1:24]
combined_ann_genes <- cbind(combined_ann_genes, ann.genes1$symbol)
colnames(combined_all_counts) <- c("S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8", "S9", "S10", "S11", "S12", "S13", "S14", "S15", "S16", "S17", "S18", "S19", "S20", "S21", "S22", "S23", "S24")
colnames(combined_ann_genes) <- c("S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8", "S9", "S10", "S11", "S12", "S13", "S14", "S15", "S16", "S17", "S18", "S19", "S20", "S21", "S22", "S23", "S24", "symbol")
combined_metadata <- metadata2
combined_metadata <- combined_metadata[1:3]
rownames(combined_metadata) <- c("S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8", "S9", "S10", "S11", "S12", "S13", "S14", "S15", "S16", "S17", "S18", "S19", "S20", "S21", "S22", "S23", "S24")
combined_metadata$pair <- c(rep(seq(1:3), each = 8))
rownames(qc1) <- c("S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8", "S9", "S10", "S11", "S12", "S13", "S14", "S15", "S16", "S17", "S18", "S19", "S20", "S21", "S22", "S23", "S24")
rownames(qc2) <- c("S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8", "S9", "S10", "S11", "S12", "S13", "S14", "S15", "S16", "S17", "S18", "S19", "S20", "S21", "S22", "S23", "S24")
rownames(qc3) <- c("S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8", "S9", "S10", "S11", "S12", "S13", "S14", "S15", "S16", "S17", "S18", "S19", "S20", "S21", "S22", "S23", "S24")
combined_qc <- qc1[,c(1:16, 20:24)] + qc2[,c(1:16, 20:24)] + qc3[,c(1:16, 20:24)]
combined_qc <- cbind(combined_qc, qc2[,c(17:19, 25:26)])
combined_summarydata <- summarydata1[, c(3:5, 15:16)] + summarydata2[, c(3:5, 15:16)] + summarydata3[, c(3:5, 15:16)]
combined_summarydata <- cbind(combined_summarydata, summarydata2[,c(25:27,29)])
rownames(combined_summarydata) <- c("S1", "S2", "S3", "S4", "S5", "S6", "S7", "S8", "S9", "S10", "S11", "S12", "S13", "S14", "S15", "S16", "S17", "S18", "S19", "S20", "S21", "S22", "S23", "S24")

get_heatmap_fn = function(combined_metadata) {
        # return the pheatmap function with or without metadata   
        if(ncol(combined_metadata) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=combined_metadata, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn = get_heatmap_fn(combined_metadata)

human = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                dataset="hsapiens_gene_ensembl",
                host = "jul2015.archive.ensembl.org")
conversions = getBM(attributes=c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),
                    mart=human)
rrna_biotypes = c("rRNA", "Mt_rRNA", "misc_RNA", "snRNA", "snoRNA",
                  "tRNA", "Mt_tRNA")

rrna_genes <- unique(subset(conversions, gene_biotype %in% rrna_biotypes)$ensembl_gene_id)

mt_genes <- c("ENSG00000210049","ENSG00000211459","ENSG00000210077","ENSG00000210082","ENSG00000209082","ENSG00000198888","ENSG00000210100","ENSG00000210107","ENSG00000210112","ENSG00000198763","ENSG00000210117","ENSG00000210127","ENSG00000210135","ENSG00000210140","ENSG00000210144","ENSG00000198804","ENSG00000210151","ENSG00000210154","ENSG00000198712","ENSG00000210156","ENSG00000228253","ENSG00000198899","ENSG00000198938","ENSG00000210164","ENSG00000198840","ENSG00000210174","ENSG00000212907","ENSG00000198886","ENSG00000210176","ENSG00000210184","ENSG00000210191","ENSG00000198786","ENSG00000198695","ENSG00000210194","ENSG00000198727","ENSG00000210195","ENSG00000210196")

#Protein coding genes
human = useMart(biomart = "ENSEMBL_MART_ENSEMBL",
                dataset="hsapiens_gene_ensembl",
                host = "jul2015.archive.ensembl.org")
protein_coding <- getBM(attributes=c("ensembl_gene_id", "hgnc_symbol", "gene_biotype"),filters = "biotype", values="protein_coding",mart=human)
pc_genes <- protein_coding[,1]

counts_pc_genes <- combined_all_counts[rownames(combined_all_counts) %in% pc_genes,]
sum_pc_genes <- colSums(counts_pc_genes, na.rm = FALSE, dims = 1)

#Reads aligning to mt genes
counts_mt_genes <- combined_all_counts[rownames(combined_all_counts) %in% mt_genes,]
sum_mt_genes <- colSums(counts_mt_genes, na.rm = FALSE, dims = 1)
#sum(sum_mt_genes)
#Reads aligning to rRNA genes
counts_rrna_genes <- combined_all_counts[rownames(combined_all_counts) %in% rrna_genes,]
sum_rrna_genes <- colSums(counts_rrna_genes, na.rm = FALSE, dims = 1)
#sum(sum_rrna_genes)
```

## Boxplot of log10 TMM-normalized counts per gene
The spread of the log10 TMM-normalized counts per gene data should be similar for every sample. Optimal data would display similar boxplots for each sample without many outliers. 

This plot shows some variation between samples for the general spread of the data.

Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r counts_density_combined}
#Analysis without mt and rRNA genes
combined_counts_no_rrna = combined_all_counts[!rownames(combined_all_counts) %in% rrna_genes & !rownames(combined_all_counts) %in% mt_genes,]
combined_anngenes_counts_no_rrna = combined_ann_genes[!rownames(combined_ann_genes) %in% rrna_genes & !rownames(combined_ann_genes) %in% mt_genes,]
y = DGEList(counts=combined_counts_no_rrna)
y = calcNormFactors(y)
normalized_combined_counts_no_rrna = cpm(y, normalized.lib.sizes=TRUE)
ann.norm.counts <- merge(normalized_combined_counts_no_rrna, ann.genes1, by=0)
row.names(ann.norm.counts) <- ann.norm.counts$Row.names
ann.norm.counts <- ann.norm.counts[,c(2:25,50)]
ann_norm_counts_no_rrna <- ann.norm.counts[!rownames(ann.norm.counts) %in% rrna_genes & !rownames(ann.norm.counts) %in% mt_genes,]
#normalized_combined_counts_no_rrna <- normalized_combined_counts_no_rrna[,c(1:12, 14:24)]

melted = melt(normalized_combined_counts_no_rrna)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
Density of normalized counts per gene should be very similar between samples, with no large shifts in any of the sample curves.

Our data shows quite a bit of variation between samples, which may skew the differential expression results. There is a large shift in a curve for a single large outlier, S13, which is likely due to the low number of mapped reads.

```{r density-normalized-combined}
ggplot(melted, aes(x=count, group=sample)) +
        geom_density() +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) + xlab("")
```

## Heatmap of marker genes
```{r Heatmap of markers_combined}
marker_norm_counts <- subset(ann_norm_counts_no_rrna, symbol == "IFNG" | symbol == "IL17A")
marker_norm_counts <- marker_norm_counts[1:24]
colnames(marker_norm_counts) <- rownames(combined_summarydata)
heatmap_fn(cor(marker_norm_counts, method="spearman"))
heatmap_fn((marker_norm_counts))


###PCA plot (no mt genes) - by cell type
#Since it's hard to determine whether the groups cluster well by PCA plot when all cell types are visualized together, clustering of samples for each individual cell type was explored.

#None of the cell types displayed a clear separation of healthy and patient samples. The large variation between samples in the same group and no clear distinction between sample groups will make identification of differentially expressed genes difficult. However, the PCA is sensitive to outliers, and the normalized values used for the PCA analysis have not been log2 transformed.

#### Cell type: DP
DP_meta <- subset(combined_metadata, celltype == "DP")
DP_norm_counts <- normalized_combined_counts_no_rrna[, row.names(DP_meta)]

pca_matrix <- prcomp(t(DP_norm_counts))$x
df <- cbind(DP_meta, pca_matrix[, c("PC1", "PC2")])

#ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
#size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
#0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")

#### Cell type: Th17
#combined_metadata <-combined_metadata[c(1:12,14:24),]  
Th17_meta <- subset(combined_metadata, celltype == "Th17")
Th17_norm_counts <- normalized_combined_counts_no_rrna[, row.names(Th17_meta)]

pca_matrix <- prcomp(t(Th17_norm_counts))$x
df <- cbind(Th17_meta, pca_matrix[, c("PC1", "PC2")])

#ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
#size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
#0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")

#### Cell type: Th1
Th1_meta <- subset(combined_metadata, celltype == "Th1")
Th1_norm_counts <- normalized_combined_counts_no_rrna[, row.names(Th1_meta)]

pca_matrix <- prcomp(t(Th1_norm_counts))$x
df <- cbind(Th1_meta, pca_matrix[, c("PC1", "PC2")])

#ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
#size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
#0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")

#### Cell type: DN
DN_meta <- subset(combined_metadata, celltype == "DN")
DN_norm_counts <- normalized_combined_counts_no_rrna[, row.names(DN_meta)]

pca_matrix <- prcomp(t(DN_norm_counts))$x
df <- cbind(DN_meta, pca_matrix[, c("PC1", "PC2")])

#ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
#size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
#0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")

## Correlation (Spearman) heatmap of TMM-normalized counts (no mt genes) - by cell type
#None of the cell types cluster by healthy and patient samples, again, indicating the large amount of variation within sample groups. However, the heatmap is also sensitive to outliers, and the normalized values used for the heatmap analysis have not been log2 transformed.

#### Cell type: DP
#heatmap_fn(cor(DP_norm_counts, method="spearman"))

#### Cell type: Th17
#heatmap_fn(cor(Th17_norm_counts, method="spearman"))

#### Cell type: Th1
#heatmap_fn(cor(Th1_norm_counts, method="spearman"))

#### Cell type: DN
#heatmap_fn(cor(DN_norm_counts, method="spearman"))

##Analysis of Genes of Interest (15 genes)
#Overall, the sample groups do not cluster for these genes, so, again, it is not likely that there will be many significant differences between sample groups for these genes.

###Analysis of Genes of Interest (15 genes) - PCA
y = DGEList(counts=combined_counts_no_rrna)
y = calcNormFactors(y)
normalized_combined_counts_no_rrna = cpm(y, normalized.lib.sizes=TRUE)
ann.norm.counts <- merge(normalized_combined_counts_no_rrna, ann.genes, by=0)
row.names(ann.norm.counts) <- ann.norm.counts$Row.names
ann.norm.counts <- ann.norm.counts[,c(2:25,50)]
ann_norm_counts_no_rrna = ann.norm.counts[!rownames(ann.norm.counts) %in% rrna_genes & !rownames(ann.norm.counts) %in% mt_genes,]

ccl3 <- subset(ann_norm_counts_no_rrna, symbol == "CCL3")
ccl4 <- subset(ann_norm_counts_no_rrna, symbol == "CCL4")
ccl5 <- subset(ann_norm_counts_no_rrna, symbol == "CCL5")
ccl6 <- subset(ann_norm_counts_no_rrna, symbol == "CSF2")
ccl7 <- subset(ann_norm_counts_no_rrna, symbol == "CXCR3")
ccl8 <- subset(ann_norm_counts_no_rrna, symbol == "EPSTI1")
ccl9 <- subset(ann_norm_counts_no_rrna, symbol == "GZMB")
ccl10 <- subset(ann_norm_counts_no_rrna, symbol == "IFNG")
ccl11<- subset(ann_norm_counts_no_rrna, symbol == "IL23R")
ccl12<- subset(ann_norm_counts_no_rrna, symbol == "STAT3")
ccl13<- subset(ann_norm_counts_no_rrna, symbol == "IL3")
ccl14<- subset(ann_norm_counts_no_rrna, symbol == "RGS2")
ccl15 <- subset(ann_norm_counts_no_rrna, symbol == "STAT1")
ccl16<- subset(ann_norm_counts_no_rrna, symbol == "STAT5A")
ccl17<- subset(ann_norm_counts_no_rrna, symbol == "TBX21")
genes <- data.frame()
genes <- rbind(ccl3, ccl4, ccl5, ccl6, ccl7, ccl8, ccl9, ccl10, ccl11, ccl12, ccl13, ccl14, ccl15, ccl16, ccl17)

row.names(genes) <- genes$symbol
genes <- genes[,1:24]
order <- c(1,2,3,4,1,2,3,4,1,2,3,4,2,1,3,4,1,2,3,4,1,2,3,4)
genes <- genes[, order(order),]

#PCA plot of 15 genes
pca_matrix <- prcomp(t(genes))$x
df <- cbind(combined_metadata, pca_matrix[, c("PC1", "PC2")])

#ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
#size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3,0.3)) + ggtitle("PC1 vs PC2 :: normalized #counts (no rrna)")   

###Analysis of Genes of Interest (15 genes) - Heatmap
#Ordered the heatmap by celltype, but no marked expression differences between the healthy and patient cells for any of the cell types.

combined_metadata_15 <- combined_metadata
combined_metadata_15 <- cbind(combined_metadata_15, order)
combined_metadata_15 <- combined_metadata_15[order(combined_metadata_15$order,combined_metadata_15$samplegroup),]
#heatmap_fn(genes, method="pearson", scale = "row", show_rownames = T,cluster_rows=F, cluster_cols=F)
#write.table(genes, file ="normalized_counts_15_genes_combined.txt", quote = FALSE, sep = "\t", row.names=T)

#Raw counts table
ccl3 <- subset(combined_ann_genes, symbol == "CCL3")
ccl4 <- subset(combined_ann_genes, symbol == "CCL4")
ccl5 <- subset(combined_ann_genes, symbol == "CCL5")
ccl6 <- subset(combined_ann_genes, symbol == "CSF2")
ccl7 <- subset(combined_ann_genes, symbol == "CXCR3")
ccl8 <- subset(combined_ann_genes, symbol == "EPSTI1")
ccl9 <- subset(combined_ann_genes, symbol == "GZMB")
ccl10 <- subset(combined_ann_genes, symbol == "IFNG")
ccl11<- subset(combined_ann_genes, symbol == "IL23R")
ccl12<- subset(combined_ann_genes, symbol == "STAT3")
ccl13<- subset(combined_ann_genes, symbol == "IL3")
ccl14<- subset(combined_ann_genes, symbol == "RGS2")
ccl15 <- subset(combined_ann_genes, symbol == "STAT1")
ccl16<- subset(combined_ann_genes, symbol == "STAT5A")
ccl17<- subset(combined_ann_genes, symbol == "TBX21")
counts_15_genes <- rbind(ccl3, ccl4, ccl5, ccl6, ccl7, ccl8, ccl9, ccl10, ccl11, ccl12, ccl13, ccl14, ccl15, ccl16, ccl17)
row.names(counts_15_genes) <- counts_15_genes$symbol
counts_15_genes <- counts_15_genes[, c(1:24)]
#write.table(counts_15_genes, "raw_counts_15_genes.txt", sep="\t", quote=F)

###Analysis of Genes of Interest (15 genes) - HDP/HTh17
#Ordered the heatmap by celltype, some of the candidate genes show obvious differences in expression, including GZMB, CCL4, and CCL5. We may look for these genes being differentially expressed between healthy DP and Th17 cells.

get_heatmap_fn_th_dp = function(summarydata_merged) {
        # return the pheatmap function with or without metadata   
        if(ncol(combined_metadata_15) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=combined_metadata_15, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn_th_dp = get_heatmap_fn_th_dp(summarydata_merged)

genes_th_dp <- genes[, c("S6", "S22", "S5", "S14", "S21")]
combined_metadata_15_th_dp <- combined_metadata_15[,-5]
#heatmap_fn_th_dp(genes_th_dp, method="pearson", scale = "row", show_rownames = T,cluster_rows=F, cluster_cols=F)
```


# Differential Expression analysis 
Differential gene expression analysis of count data was performed using the Bioconductor R package, [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html). The count data was fit to a negative binomial model and dispersion estimates were generated using the mean values from the maximum likelikhood estimate of log2 fold changes, optimizing the Cox-Reid adjusted profile likelihood. 

Several quality metrics were assessed to explore the fit of the count data to each of the four cell type models, and differential expression analysis was performed for each cell type individually. The models used to estimate differential expression accounted for the matched sample design.

For all cell types, the data was very noisy, and the models were not able to account for all variation in the data. Since the fitting of the data to the models was not good, you must be very careful how you interpret the results. Significantly differentially expressed genes identified in these analyses will require exhaustive lab verification. In addition, we recommend looking at the expression heatmaps of the significant DE genes to narrow down genes most likely to be truly differentially expressed between healthy individuals and patients. 

Also, be aware that some of the significant DE genes are related to expression of rRNA, which could be an artifact of your experimental protocol resulting in an excess of rRNA / mtRNA in your samples.

The raw counts file below can be used to perform your own differential expression analysis. The normalized counts file should be used for any sample comparisons or plots.

```{r normalized_counts_file}
#write.table(combined_anngenes_counts_no_rrna, "raw_counts_no_rrna_genes.txt", sep="\t", quote=F)
```
[Download raw counts file used for all DE Analyses (no rrna/mtrna genes)](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/raw_counts_no_rrna_genes.txt)

```{r de-setup_normalized counts}
library(DESeq2)
library(vsn)

design = ~ pair + celltype + disease
condition = "disease"

# Differential expression design parameters
total_counts <- combined_counts_no_rrna[, row.names(combined_metadata)]

dds = DESeqDataSetFromMatrix(countData=total_counts,
                             colData=combined_metadata, design = design)
dds = DESeq(dds)

total_norm_counts <- counts(dds, normalized=T)
total_norm_counts <- data.frame(total_norm_counts)
total_norm_counts$names <- rownames(total_norm_counts) 

combined_anngenes_counts_no_rrna_names <- combined_anngenes_counts_no_rrna
combined_anngenes_counts_no_rrna_names$names <- rownames(combined_anngenes_counts_no_rrna)
norm_counts <- merge(total_norm_counts, combined_anngenes_counts_no_rrna_names, by="names")
rownames(norm_counts) <- norm_counts$names
norm_counts <- norm_counts[,c(2:25,50)]
names(norm_counts) <- c(rownames(combined_metadata), "symbol")
#write.table(norm_counts, "norm_counts_no_rrna_genes.txt", sep="\t", quote=F)
```
[Download normalized counts file (no rrna/mtrna genes)](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/norm_counts_no_rrna_genes.txt)

## Differential Expression analysis - Healthy DP versus Healthy Th17
To determine whether we can detect differential expression of the candidate genes identified using the NanoStream we explored the differential expression between healthy DP cells and healthy Th17 cells. Overall, we were unable to identify significant differences in expression of the candidate genes other than interferon gamma. The variation within samples allows only for the identification of DEGs exhibiting large fold changes in expression levels.

```{r de-setup_combined_th_dp}
library(DESeq2)
library(vsn)

design = ~ pair + celltype
condition = "celltype"

# Differential expression design parameters

th_dp_meta <- subset(combined_metadata, samplegroup == "HTh17" | samplegroup == "HDP")
th_dp_meta <- th_dp_meta[-3,]
th_dp_meta$celltype <- factor(th_dp_meta$celltype)
th_dp_meta$disease <- factor(th_dp_meta$disease)
th_dp_meta$pair <- factor(th_dp_meta$pair)
th_dp_counts <- combined_counts_no_rrna[, row.names(th_dp_meta)]

counts <- th_dp_counts[rowSums(th_dp_counts>0)>1,]
dds = DESeqDataSetFromMatrix(countData=counts,
                             colData=th_dp_meta, design = design)
dds = DESeq(dds)

contrast2 <- list("celltypeDP", "celltypeTh17")
th_dp <- results(dds, contrast=contrast2)
#summary(dp_th)
```

### Effect of variance stabilization - DP versus Th17
For RNA-Seq gene counts, the variance increases with the mean. To account for this variance, logarithmic transformation (log2) of normalized count values will ensure that these genes won't dominate the PCA plots. However, due to the noise associated with low count values, the general log2 transformation will worsen this noise, and low count genes will instead dominate the PCA plots. Therefore, we need to use a transformation that will stabilize the variance across the mean for the gene counts. The plots below show the standard deviation of transformed counts using log2, rlog, and vsd transformations by rank(mean) (from top to bottom, respectively). The transformations are unable to greatly reduce the standard deviation, which is in line with there being a large variation between replicates associated with this dataset. 

```{r deseq-diagnostics_th_dp}
#For RNA-Seq raw counts, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and show the low count genes will dominate the PCA plots. Therfore, transform to stabilize variance across the mean using rlog. For high counts, gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

#Plotting standard deviation by rank(mean)
meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

### Similarity between samples - DP versus Th17
Since the rlog transformation best minimizes variance and standard deviation, we will use the rlog transformation to estimate variation between samples with Euclidean distance and PCA. 

### Dispersion estimates - DP versus Th17
The following plot shows the dispersion by mean of normalized counts. The dispersion does not look good, since we expect the dispersion to decrease with increased mean of normalized counts. The dispersion here does not show much of a trend and appears more scattered.

```{r dispersion-estimate_th_dp}
plotDispEsts(dds)
```

### MA-plots - DP versus Th17
MA plots explore the mean expression level of the genes with the fold change, allowing visualization of the genes that are differentially expressed (red). Due to the large degree of variance within our sample groups, we are only able to detect genes with high expression and large fold changes between conditions.

```{r DESeq-output_MA_th_dp}
#res <- results(dds, alpha=.05)
plotMA(th_dp, ylim=c(-5,5))
```
### Volcano-plot - DP versus Th17
The following volcano plot visualizes the significant DE genes. Due to the variation in the data within the celltype groups, we are unable to detect any significant differential expression between healthy and diseased patients for genes other than those with large fold changes. 
```{r DESeq-volcano_th_dp}

df <- data.frame(th_dp)
# ggplot(data=df, aes(x=log2FoldChange, y=-log10(padj))) +
#   scale_color_manual(values = c("grey", "purple")) +
#   xlim(c(-2,2)) +
#   ylim(c(0,1.3)) +
#   geom_point(alpha=0.75, pch=16) +
#   theme(legend.position = "none",
#         plot.title = element_text(size = rel(1.5)),
#         axis.title = element_text(size = rel(1.5)),
#         axis.text = element_text(size = rel(1.25))) +
#   xlab("log2 fold change") + ylab("-log10 p-value")

with(th_dp, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-8,8), ylim=c(0,8)))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(th_dp, padj<.05 ), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
with(subset(th_dp, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="green"))
```

### Differentially expressed genes - DP versus Th17
Using an padj cut-off of 0.05 (p-values multiple test corrected using the BH method), 94 genes were identified as significant.

#### Significant differentially expressed genes - DP versus Th17
The list of significant DE genes contains the gene expression differences in the healthy DP cells relative to the healthy Th17 cells. The row names in the table are the Ensembl gene ids, followed by the columns: the mean of the normalized counts for that gene for all samples (`baseMean`), log2 fold change (`log2FoldChange`), standard error (`lfcSE`), Wald statistic (`stat`), Wald test p-value (`pvalue`), BH adjusted p-values, which account for multiple testing (`padj`), and the official gene symbol (`symbol`). 

For example, in the first row of the table for gene id ENSG00000104325 (official gene symbol = DECR1), the DP healthy cells had an expression level log2 fold change of 8.479 relative to the Th17 healthy cells, with an average mean expression of 338.58 (norm. counts), and the gene was up-regulated in healthy DP cells relative to healthy Th17 cells, since the log2FoldChange is positive.

```{r DE_genes_list_th_dp}
#Gene list with padj
resSig <- subset(th_dp, padj < 0.05)
DEG <- data.frame(baseMean = resSig[[1]], log2FoldChange = resSig[[2]], lfcSE = resSig[[3]], stat = resSig[[4]], pvalue = resSig[[5]], padj = resSig[[6]], row.names = row.names(resSig))
DEG <- cbind(DEG, name = row.names(DEG))
DEGsymbol <- which(row.names(ann.genes1) %in% row.names(DEG))
DEGsymbol <- ann.genes1[DEGsymbol,]
DEGsymbol <- cbind(DEGsymbol, name = row.names(DEGsymbol))
DEG_genenames_th_dp <- merge(DEG, DEGsymbol, by="name")

th_dp_norm_counts <- normalized_combined_counts_no_rrna[, row.names(th_dp_meta)]
th_dp_sig_norm_counts <- cbind(th_dp_norm_counts, "name" = row.names(th_dp_norm_counts))
th_dp_sig_norm_counts <- merge(th_dp_sig_norm_counts, DEG_genenames_th_dp, by="name")
row.names(th_dp_sig_norm_counts) <- th_dp_sig_norm_counts$name
th_dp_sig_norm_counts <- th_dp_sig_norm_counts[, c(2:6,37)]

row.names(DEG_genenames_th_dp) <- DEG_genenames_th_dp$name
DEG_genenames_th_dp <- DEG_genenames_th_dp[, c(2:7,32)]
DEG_genenames_th_dp <- DEG_genenames_th_dp[order(DEG_genenames_th_dp$padj),]
knitr::kable(DEG_genenames_th_dp)
#write.table(DEG_genenames_th_dp, "th_dp_results", sep="\t", quote=F)

#Normalized counts of genes
DEG_genenames_th_dp <- merge(DEG, DEGsymbol, by="name")
th_dp_sig_counts <- cbind(th_dp_counts, "name" = row.names(th_dp_counts))
th_dp_sig_counts <- merge(th_dp_sig_counts, DEG_genenames_th_dp, by="name")
row.names(th_dp_sig_counts) <- th_dp_sig_counts$name
th_dp_sig_counts <- th_dp_sig_counts[, c(2:6)]

log2_counts <- assay(rld[notAllZero,])
log2_counts <- as.data.frame(log2_counts) 
idx <- rownames(log2_counts) %in% rownames(th_dp_sig_counts)
log2_sig_counts <- log2_counts[idx,]
log2_sig_counts$name <- rownames(log2_sig_counts)
log2_sig_counts_symbol <- merge(log2_sig_counts, DEG_genenames_th_dp, by="name")
rownames(log2_sig_counts_symbol) <- log2_sig_counts$name 
log2_sig_counts_symbol <- log2_sig_counts_symbol[, c(2:6,37)]
#write.table(log2_sig_counts_symbol, "th_dp_log2_counts", sep="\t", quote=F)
```

[Download th_dp Results](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/th_dp_results)

#### Significant differentially expressed genes - DP versus Th17
This plot shows the expression of the significant differentially expressed genes by sample. These genes do cluster by sample group; however, there is a lot of variability in expression of these genes across sample group. Lab validation would be extremely important for any of the identified significant genes. The scale values compare the sample expression of a specific gene to the mean expression of the gene across all samples.
```{r heatmap_DE_genes_th_dp}
get_heatmap_fn = function(summarydata_merged) {
        # return the pheatmap function with or without metadata
        if(ncol(th_dp_meta) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=th_dp_meta, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn = get_heatmap_fn(summarydata_merged)

DEG_genenames_th_dp <- merge(DEG, DEGsymbol, by="name")
th_dp_sig_counts <- cbind(th_dp_counts, "name" = row.names(th_dp_counts))
th_dp_sig_counts <- merge(th_dp_sig_counts, DEG_genenames_th_dp, by="name")
row.names(th_dp_sig_counts) <- th_dp_sig_counts$name
th_dp_sig_counts <- th_dp_sig_counts[, c(2:6)]

log2_counts <- assay(rld[notAllZero,])
log2_counts <- as.data.frame(log2_counts) 
idx <- rownames(log2_counts) %in% rownames(th_dp_sig_counts)
log2_sig_counts <- log2_counts[idx,]

#Values relate to the row mean subtracted from the normalized count value for each sample value.
heatmap_fn(log2_sig_counts, clustering_method = "ward.D2", clustering_distance_cols = "correlation")
```
[Download th_dp rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/th_dp_log2_counts)

### Genes of Interest -  Disease vs. Healthy subset - DP versus Th17
The following tables and plots explore the counts and expression patterns for the genes of interest.

#### Genes of Interest -- results table and counts - DP versus Th17
Among the genes of interest, none exhibited significant differential expression between the DP and Th17 cells except Interferon gamma (IFNG). However, due to the high variation within samples, it is likely that there are many differentially expressed genes that we were unable to identify. We were unable to identify DEGs that didn't have huge fold changes between samples, so even though some genes appeared to be expressed at different levels between groups (GZMB, CCL4, CCL5), they were not found to be significant.

```{r candidate_gene_table_th_dp}
genes <- data.frame(baseMean = th_dp[[1]], log2FoldChange = th_dp[[2]], lfcSE = th_dp[[3]], stat = th_dp[[4]], pvalue = th_dp[[5]], padj = th_dp[[6]], row.names = row.names(th_dp))
genes <- cbind(genes, name = row.names(genes))
genessymbol <- which(row.names(ann.genes1) %in% row.names(genes))
genessymbol <- ann.genes1[genessymbol,]
genessymbol <- cbind(genessymbol, name = row.names(genessymbol))
genes_genenames <- merge(genes, genessymbol, by="name")
row.names(genes_genenames) <- genes_genenames$name
genes_genenames <- genes_genenames[, c(2:7,32)]
CCL3 <- genes_genenames[which(genes_genenames$symbol=="CCL3"),]
CCL4 <- genes_genenames[which(genes_genenames$symbol=="CCL4"),]
CCL5 <- genes_genenames[which(genes_genenames$symbol=="CCL5"),]
CSF2 <- genes_genenames[which(genes_genenames$symbol=="CSF2"),]
CXCR3 <- genes_genenames[which(genes_genenames$symbol=="CXCR3"),]
EPSTI1 <- genes_genenames[which(genes_genenames$symbol=="EPSTI1"),]
GZMB <- genes_genenames[which(genes_genenames$symbol=="GZMB"),]
IFNG <- genes_genenames[which(genes_genenames$symbol=="IFNG"),]
IL23R <- genes_genenames[which(genes_genenames$symbol=="IL23R"),]
IL3 <- genes_genenames[which(genes_genenames$symbol=="IL3"),]
RGS2 <- genes_genenames[which(genes_genenames$symbol=="RGS2"),]
STAT1 <- genes_genenames[which(genes_genenames$symbol=="STAT1"),]
STAT3 <- genes_genenames[which(genes_genenames$symbol=="STAT3"),]
STAT5A <- genes_genenames[which(genes_genenames$symbol=="STAT5A"),]
TBX21 <- genes_genenames[which(genes_genenames$symbol=="TBX21"),]
marker_genes <- rbind(CCL3, CCL4, CCL5, CSF2, CXCR3, EPSTI1, GZMB, IFNG, IL23R, IL3, RGS2, STAT1, STAT3, STAT5A, TBX21)

marker_counts_th_dp <- as.data.frame(log2_counts[row.names(marker_genes),])
row.names(marker_counts_th_dp) <- marker_genes$symbol
marker_genes <- marker_genes[,1:7]
knitr::kable(marker_genes)
#write.table(marker_genes, "th_dp_marker_counts", sep="\t", quote=F)
```

#### Genes of Interest -- expression patterns - DP versus Th17
Overall, the genes of interest did not cluster well by cell type. Differences in gene expression between sample groups is observed only for interferon gamma, CCL4, and CCL5; however, based on our DE analysis, the difference is not significant for CCL4 and CCL5.

```{r candidate_gene_heatmap_th_dp}

#Values relate to the row mean subtracted from the normalized count value for each sample value.
idx <- rownames(log2_counts) %in% rownames(marker_genes)
log2_marker_counts_th_dp <- log2_counts[idx,]
log2_marker_counts_th_dp$name <- rownames(log2_marker_counts_th_dp)
log2_marker_counts_th_dp_symbol <- merge(log2_marker_counts_th_dp, genessymbol, by="name")
row.names(log2_marker_counts_th_dp_symbol) <- log2_marker_counts_th_dp_symbol$symbol
log2_marker_counts_th_dp_symbol <- log2_marker_counts_th_dp_symbol[, c(2:6)]
heatmap_fn(log2_marker_counts_th_dp_symbol, clustering_method = "ward.D2", clustering_distance_cols = "correlation")
```
[Download Marker Genes - rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/th_dp_marker_counts)

### Functional Analysis - DP versus Th17
Using the DE genes, a list of statistically enriched gene ontology (GO) terms was generated using the program [gprofileR](http://biit.cs.ut.ee/gprofiler/). From the single significant GO term associated with the 94 DE genes, we can see that enrichment of gene ontology terms generates little information regarding affected pathways.
```{r functional_analysis_th_dp}
gene_list <- DEG_genenames_th_dp
gene_list <- cbind(gene_id = gene_list$symbol, gene_list)
gene_list <- gene_list[, c(1,7)]

#gprofileR
library(gProfileR)
gprofiler_results <- gprofiler(query = gene_list, organism = "hsapiens", ordered_query = F, 
                               exclude_iea = F, max_set_size = 0, correction_method = "fdr", 
                               hier_filtering = "none", domain_size = "annotated", custom_bg = "")

allterms <- gprofiler_results$term.id
GOs <- allterms[grep("GO:", allterms)]
pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
GO.pval <- cbind(gprofiler_results$term.id, gprofiler_results$p.value)
GO_genes_th_dp <- gprofiler_results[, c("term.id", "term.name", "p.value", "term.size", "overlap.size", "intersection")]
names(GO_genes_th_dp) <- c("term.id", "term.name", "p.value", "term.size", "overlap.size", "assoc.gene.ids")
#write.table(GO_genes_th_dp, file ="GO_genes_th_dp", quote = FALSE, sep = "\t", row.names=T)
```
[Download GO Terms](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/GO_genes_th_dp)


## Differential Expression analysis - Disease vs. Healthy subset - DP cells
The following analysis performs differential expression testing for healthy controls versus patients in DP cells.
```{r de-setup_combined_DP}
library(DESeq2)
library(vsn)

design = ~ pair + disease
condition = "disease"

# Differential expression design parameters

DP_meta <- subset(combined_metadata, celltype == "DP")
DP_meta$disease <- factor(DP_meta$disease)
DP_meta$pair <- factor(DP_meta$pair)
DP_counts <- combined_counts_no_rrna[, row.names(DP_meta)]

counts <- DP_counts[rowSums(DP_counts>0)>1,]
dds = DESeqDataSetFromMatrix(countData=counts,
                             colData=DP_meta, design = design)
dds = DESeq(dds)

contrast2 <- list("diseasehealthy", "diseasepatient")
DP <- results(dds, contrast=contrast2)
#summary(DP)
```

### Effect of variance stabilization
For RNA-Seq gene counts, the variance increases with the mean. To account for this variance, logarithmic transformation (log2) of normalized count values will ensure that these genes won't dominate the PCA plots. However, due to the noise associated with low count values, the general log2 transformation will worsen this noise, and low count genes will instead dominate the PCA plots. Therefore, we need to use a transformation that will stabilize the variance across the mean for the gene counts. The plots below show the standard deviation of transformed counts using log2, rlog, and vsd transformations by rank(mean) (from top to bottom, respectively). The transformations are unable to greatly reduce the standard deviation, which is in line with there being a large variation between replicates associated with this dataset. 

```{r deseq-diagnostics_DP}
#For RNA-Seq raw counts, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and show the low count genes will dominate the PCA plots. Therfore, transform to stabilize variance across the mean using rlog. For high counts, gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

#Plotting standard deviation by rank(mean)
meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

### Similarity between samples
Since the rlog transformation best minimizes variance and standard deviation, we will use the rlog transformation to estimate variation between samples with Euclidean distance and PCA. 

### Dispersion estimates
The following plot shows the dispersion by mean of normalized counts. The dispersion does not look good, since we expect the dispersion to decrease with increased mean of normalized counts. The dispersion here does not show much of a trend and appears more scattered.

```{r dispersion-estimate_DP}
plotDispEsts(dds)
```

### MA-plots
MA plots explore the mean expression level of the genes with the fold change, allowing visualization of the genes that are differentially expressed (red). Due to the large degree of variance within our sample groups, we are only able to detect genes with high expression and large fold changes between conditions.

```{r DESeq-output_MA_DP}
#res <- results(dds, alpha=.05)
plotMA(DP, ylim=c(-5,5))
```
### Volcano-plot
The following volcano plot visualizes the significant DE genes. Due to the variation in the data within the DP group, we are unable to detect any significant differential expression between healthy and diseased patients in DP cells for genes other than those with large fold changes. 
```{r DESeq-volcano_DP}

df <- data.frame(DP)
# ggplot(data=df, aes(x=log2FoldChange, y=-log10(padj))) +
#   scale_color_manual(values = c("grey", "purple")) +
#   xlim(c(-2,2)) +
#   ylim(c(0,1.3)) +
#   geom_point(alpha=0.75, pch=16) +
#   theme(legend.position = "none",
#         plot.title = element_text(size = rel(1.5)),
#         axis.title = element_text(size = rel(1.5)),
#         axis.text = element_text(size = rel(1.25))) +
#   xlab("log2 fold change") + ylab("-log10 p-value")

with(DP, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-8,8), ylim=c(0,8)))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(DP, padj<.05 ), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
with(subset(DP, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="green"))
```

### Differentially expressed genes
Using an padj cut-off of 0.05 (p-values multiple test corrected using the BH method), 20 genes were identified as significant.

#### Significant differentially expressed genes - Disease vs. Healthy subset - gene list - DP cells
The list of significant DE genes contains the gene expression differences in the patients  relative to the healthy controls. The row names in the table are the Ensembl gene ids, followed by the columns: the mean of the normalized counts for that gene for all samples (`baseMean`), log2 fold change (`log2FoldChange`), standard error (`lfcSE`), Wald statistic (`stat`), Wald test p-value (`pvalue`), BH adjusted p-values (`padj`), and the official gene symbol (`symbol`). 

For example, in the first row of the table for gene id ENSG00000249055 (official gene symbol = TBCAP3), the DP cells of the patients had an expression level log2 fold change of 6.73 relative to the DP cells of the healthy controls, with an average mean expression of 423.97 (norm. counts), and the gene was up-regulated in patients relative to healthy controls, since the log2FoldChange is positive.

The short list of significantly differentially expressed genes is likely due to the high variability seen within sample groups.

```{r DE_genes_list_DP}
#Gene list with padj
resSig <- subset(DP, padj < 0.05)
DEG <- data.frame(baseMean = resSig[[1]], log2FoldChange = resSig[[2]], lfcSE = resSig[[3]], stat = resSig[[4]], pvalue = resSig[[5]], padj = resSig[[6]], row.names = row.names(resSig))
DEG <- cbind(DEG, name = row.names(DEG))
DEGsymbol <- which(row.names(ann.genes1) %in% row.names(DEG))
DEGsymbol <- ann.genes1[DEGsymbol,]
DEGsymbol <- cbind(DEGsymbol, name = row.names(DEGsymbol))
DEG_genenames_DP <- merge(DEG, DEGsymbol, by="name")

DP_sig_norm_counts <- cbind(DP_norm_counts, "name" = row.names(DP_norm_counts))
DP_sig_norm_counts <- merge(DP_sig_norm_counts, DEG_genenames_DP, by="name")
row.names(DP_sig_norm_counts) <- DP_sig_norm_counts$name
DP_sig_norm_counts <- DP_sig_norm_counts[, c(2:7,38)]

row.names(DEG_genenames_DP) <- DEG_genenames_DP$name
DEG_genenames_DP <- DEG_genenames_DP[, c(2:7,32)]
DEG_genenames_DP <- DEG_genenames_DP[order(DEG_genenames_DP$padj),]
knitr::kable(DEG_genenames_DP)
#write.table(DEG_genenames_DP, "DP_results", sep="\t", quote=F)

#Normalized counts of genes
DEG_genenames_DP <- merge(DEG, DEGsymbol, by="name")
DP_sig_counts <- cbind(DP_counts, "name" = row.names(DP_counts))
DP_sig_counts <- merge(DP_sig_counts, DEG_genenames_DP, by="name")
row.names(DP_sig_counts) <- DP_sig_counts$name
DP_sig_counts <- DP_sig_counts[, c(2:7)]

log2_counts <- assay(rld[notAllZero,])
log2_counts <- as.data.frame(log2_counts) 
idx <- rownames(log2_counts) %in% rownames(DP_sig_counts)
log2_sig_counts <- log2_counts[idx,]
log2_sig_counts$name <- rownames(log2_sig_counts)
log2_sig_counts_symbol <- merge(log2_sig_counts, DEG_genenames_DP, by="name")
rownames(log2_sig_counts_symbol) <- log2_sig_counts$name 
log2_sig_counts_symbol <- log2_sig_counts_symbol[, c(2:7,38)]
#write.table(log2_sig_counts_symbol, "DP_log2_counts", sep="\t", quote=F)
```
[Download DP Results](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/DP_results)

#### Significant differentially expressed genes - Disease vs. Healthy subset - expression patterns -DP cells
This plot shows the expression of the significant differentially expressed genes by sample. These genes do cluster by sample group, but there is a lot of variability in expression of these genes across sample group. Lab validation would be extremely important for any of the identified significant genes. The scale values compare the sample expression of a specific gene to the mean expression of the gene across all samples.
```{r heatmap_DE_genes_DP}
get_heatmap_fn = function(summarydata_merged) {
        # return the pheatmap function with or without metadata
        if(ncol(DP_meta) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=DP_meta, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn = get_heatmap_fn(summarydata_merged)

log2_sig_counts <- log2_sig_counts[,1:6]
heatmap_fn(log2_sig_counts, clustering_method = "ward.D2", clustering_distance_cols = "correlation")

pca_matrix <- prcomp(t(log2_sig_counts))$x
df <- cbind(DP_meta, pca_matrix[, c("PC1", "PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")
```
[Download DP rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/DP_log2_counts)

### Genes of Interest -  Disease vs. Healthy subset - DP
The following tables and plots explore the counts and expression patterns for the genes of interest.

#### Genes of Interest -- results table and counts - DP
Among the genes of interest, none exhibited significant differential expression in DP cells. However, due to the high variation within samples, it is likely that there are many differentially expressed genes that we were unable to identify. 

```{r candidate_gene_table_DP}
genes <- data.frame(baseMean = DP[[1]], log2FoldChange = DP[[2]], lfcSE = DP[[3]], stat = DP[[4]], pvalue = DP[[5]], padj = DP[[6]], row.names = row.names(DP))
genes <- cbind(genes, name = row.names(genes))
genessymbol <- which(row.names(ann.genes1) %in% row.names(genes))
genessymbol <- ann.genes1[genessymbol,]
genessymbol <- cbind(genessymbol, name = row.names(genessymbol))
genes_genenames <- merge(genes, genessymbol, by="name")
row.names(genes_genenames) <- genes_genenames$name
genes_genenames <- genes_genenames[, c(2:7,32)]
CCL3 <- genes_genenames[which(genes_genenames$symbol=="CCL3"),]
CCL4 <- genes_genenames[which(genes_genenames$symbol=="CCL4"),]
CCL5 <- genes_genenames[which(genes_genenames$symbol=="CCL5"),]
CSF2 <- genes_genenames[which(genes_genenames$symbol=="CSF2"),]
CXCR3 <- genes_genenames[which(genes_genenames$symbol=="CXCR3"),]
EPSTI1 <- genes_genenames[which(genes_genenames$symbol=="EPSTI1"),]
GZMB <- genes_genenames[which(genes_genenames$symbol=="GZMB"),]
IFNG <- genes_genenames[which(genes_genenames$symbol=="IFNG"),]
IL23R <- genes_genenames[which(genes_genenames$symbol=="IL23R"),]
IL3 <- genes_genenames[which(genes_genenames$symbol=="IL3"),]
RGS2 <- genes_genenames[which(genes_genenames$symbol=="RGS2"),]
STAT1 <- genes_genenames[which(genes_genenames$symbol=="STAT1"),]
STAT3 <- genes_genenames[which(genes_genenames$symbol=="STAT3"),]
STAT5A <- genes_genenames[which(genes_genenames$symbol=="STAT5A"),]
TBX21 <- genes_genenames[which(genes_genenames$symbol=="TBX21"),]
marker_genes <- rbind(CCL3, CCL4, CCL5, CSF2, CXCR3, EPSTI1, GZMB, IFNG, IL23R, IL3, RGS2, STAT1, STAT3, STAT5A, TBX21)

marker_log2counts_DP <- as.data.frame(log2_counts[row.names(marker_genes),])
row.names(marker_log2counts_DP) <- marker_genes$symbol
marker_genes <- marker_genes[,1:7]
knitr::kable(marker_genes)
#write.table(marker_log2counts_DP, "DP_marker_counts", sep="\t", quote=F)
```

#### Genes of Interest -- expression patterns - DP
The genes of interest did not cluster well by sample group for DP cells.

```{r candidate_gene_heatmap_DP}

#Values relate to the row mean subtracted from the normalized count value for each sample value.
idx <- rownames(log2_counts) %in% rownames(marker_genes)
log2_marker_counts_DP <- log2_counts[idx,]
log2_marker_counts_DP$name <- rownames(log2_marker_counts_DP)
marker_genes$name <- rownames(marker_genes)
log2_marker_counts_DP_symbol <- merge(log2_marker_counts_DP, marker_genes, by="name")
row.names(log2_marker_counts_DP_symbol) <- log2_marker_counts_DP_symbol$symbol
log2_marker_counts_DP_symbol <- log2_marker_counts_DP_symbol[, c(2:7)]
heatmap_fn(log2_marker_counts_DP_symbol, clustering_method = "ward.D2", clustering_distance_cols = "correlation")
```
[Download Marker Genes - rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/DP_marker_counts)

### Functional Analysis - DP cells
Using the DE genes, a list of statistically enriched gene ontology (GO) terms was generated using the program [gprofileR](http://biit.cs.ut.ee/gprofiler/). From the two significant GO terms associated with the 20 DE genes, we can see that enrichment of gene ontology terms generates little information regarding affected pathways.
```{r functional_analysis_DP}
gene_list <- DEG_genenames_DP
gene_list <- cbind(gene_id = gene_list$symbol, gene_list)
gene_list <- gene_list[, c(1,7)]

#gprofileR
library(gProfileR)
gprofiler_results <- gprofiler(query = gene_list, organism = "hsapiens", ordered_query = F, 
                               exclude_iea = F, max_set_size = 0, correction_method = "fdr", 
                               hier_filtering = "none", domain_size = "annotated", custom_bg = "")

allterms <- gprofiler_results$term.id
GOs <- allterms[grep("GO:", allterms)]
pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
GO.pval <- cbind(gprofiler_results$term.id, gprofiler_results$p.value)
GO_genes_DP <- gprofiler_results[, c("term.id", "term.name", "p.value", "term.size", "overlap.size", "intersection")]
names(GO_genes_DP) <- c("term.id", "term.name", "p.value", "term.size", "overlap.size", "assoc.gene.ids")
write.table(GO_genes_DP, file ="GO_genes_DP", quote = FALSE, sep = "\t", row.names=T)
```
[Download GO Terms](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/GO_genes_DP)

## Differential Expression analysis - Disease vs. Healthy - Th17 cells
The following analysis performs differential expression testing for healthy controls versus patients in Th17 cells.
```{r de-setup_combined_Th17}
library(DESeq2)
library(vsn)
design = ~ pair + disease
condition = "disease"

# Differential expression design parameters

Th17_meta <- subset(combined_metadata, celltype == "Th17")
Th17_meta <- Th17_meta[-4, ]
Th17_meta$disease <- factor(Th17_meta$disease)
Th17_meta$pair <- factor(Th17_meta$pair)
Th17_counts <- combined_counts_no_rrna[, row.names(Th17_meta)]

counts <- Th17_counts[rowSums(Th17_counts>0)>1,]
dds = DESeqDataSetFromMatrix(countData=counts,
                             colData=Th17_meta, design = design)
dds = DESeq(dds)

contrast2 <- list("diseasehealthy", "diseasepatient")
Th17 <- results(dds, contrast=contrast2)
#summary(Th17)
```

### Effect of variance stabilization
For RNA-Seq gene counts, the variance increases with the mean. To account for this variance, logarithmic transformation (log2) of normalized count values will ensure that these genes won't dominate the PCA plots. However, due to the noise associated with low count values, the general log2 transformation will worsen this noise, and low count genes will instead the PCA plots. Therefore, we need to use a transformation that will stabilize the variance across the mean for the gene counts. The plots below show the standard deviation of transformed counts using log2, rlog, and vsd transformations by rank(mean)(from left to right, respectively). The rlog transformation (middle) minimizes the variance best for the low and high counts; however, there is still a large variance associated with this data. 

```{r deseq-diagnostics_Th17}
#For RNA-Seq raw counts, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and show the low count genes will dominate the PCA plots. Therfore, transform to stabilize variance across the mean using rlog. For high counts, gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

#Plotting standard deviation by rank(mean)
meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

### Dispersion estimates
The following plot shows the dispersion by mean of normalized counts. The dispersion does not look good, since we expect the dispersion to decrease with increased mean of normalized counts. The dispersion here does not show a strong trend and appears more scattered.

```{r dispersion-estimate_Th17}
plotDispEsts(dds)
```

### MA-plots
MA plots explore the mean expression level of the genes with the fold change, allowing visualization of the genes that are differentially expressed (red). Due to the large degree of variance within our sample groups, we are only able to detect genes with high expression and large fold changes between conditions.

```{r DESeq-MA_Th17}
#res <- results(dds, alpha=.05)
plotMA(Th17, ylim=c(-5,5))
```

### Volcano-plot
The following volcano plot visualizes the significant DE genes. Due to the variation in the data within the Th17 group, we are unable to detect any significant differential expression between healthy controls and patients in Th17 cells for genes other than those with large fold changes. 
```{r DESeq-volcano_Th17}

df <- data.frame(Th17)
# ggplot(data=df, aes(x=log2FoldChange, y=-log10(padj))) +
#   scale_color_manual(values = c("grey", "purple")) +
#   xlim(c(-2,2)) +
#   ylim(c(0,1.3)) +
#   geom_point(alpha=0.75, pch=16) +
#   theme(legend.position = "none",
#         plot.title = element_text(size = rel(1.5)),
#         axis.title = element_text(size = rel(1.5)),
#         axis.text = element_text(size = rel(1.25))) +
#   xlab("log2 fold change") + ylab("-log10 p-value")

with(Th17, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-8,8), ylim=c(0,8)))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(Th17, padj<.05 ), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
with(subset(Th17, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="green"))
```

### Differentially expressed genes
Using an padj cut-off of 0.05 (p-values multiple test corrected using the BH method), 124 genes were identified as significant. The S13 sample was identified as an outlier in the previous QC steps and removed prior to DE analysis for the Th17 celltype.

#### Significant differentially expressed genes - Disease vs. Healthy subset - gene list - Th17 cells
The list of significant DE genes contains the gene expression differences in the patients  relative to the healthy controls. The row names in the table are the Ensembl gene ids, followed by the columns: the mean of the normalized counts for that gene for all samples (`baseMean`), log2 fold change (`log2FoldChange`), standard error (`lfcSE`), Wald statistic (`stat`), Wald test p-value (`pvalue`), BH adjusted p-values (`padj`), and the official gene symbol (`symbol`). 

The short list of significantly differentially expressed genes is likely due to the high variability seen within sample groups.

```{r DE_genes_list_Th17}
#Gene list with padj
resSig <- subset(Th17, padj < 0.05)
DEG <- data.frame(baseMean = resSig[[1]], log2FoldChange = resSig[[2]], lfcSE = resSig[[3]], stat = resSig[[4]], pvalue = resSig[[5]], padj = resSig[[6]], row.names = row.names(resSig))
DEG <- cbind(DEG, name = row.names(DEG))
DEGsymbol <- which(row.names(ann.genes1) %in% row.names(DEG))
DEGsymbol <- ann.genes1[DEGsymbol,]
DEGsymbol <- cbind(DEGsymbol, name = row.names(DEGsymbol))
DEG_genenames_Th17 <- merge(DEG, DEGsymbol, by="name")

Th17_sig_norm_counts <- cbind(Th17_norm_counts[,-4], "name" = row.names(Th17_norm_counts))
Th17_sig_norm_counts <- merge(Th17_sig_norm_counts, DEG_genenames_Th17, by="name")
row.names(Th17_sig_norm_counts) <- Th17_sig_norm_counts$name
Th17_sig_norm_counts <- Th17_sig_norm_counts[, c(2:6,37)]

row.names(DEG_genenames_Th17) <- DEG_genenames_Th17$name
DEG_genenames_Th17 <- DEG_genenames_Th17[, c(2:7,32)]
DEG_genenames_Th17 <- DEG_genenames_Th17[order(DEG_genenames_Th17$padj),]
knitr::kable(DEG_genenames_Th17)
#write.table(DEG_genenames_Th17, "Th17_results", sep="\t", quote=F)

#Normalized counts of genes
DEG_genenames_Th17 <- merge(DEG, DEGsymbol, by="name")
Th17_sig_counts <- cbind(Th17_counts, "name" = row.names(Th17_counts))
Th17_sig_counts <- merge(Th17_sig_counts, DEG_genenames_Th17, by="name")
row.names(Th17_sig_counts) <- Th17_sig_counts$name
Th17_sig_counts <- Th17_sig_counts[, c(2:6)]

log2_counts <- assay(rld[notAllZero,])
log2_counts <- as.data.frame(log2_counts) 
idx <- rownames(log2_counts) %in% rownames(Th17_sig_counts)
log2_sig_counts <- log2_counts[idx,]
log2_sig_counts$name <- rownames(log2_sig_counts)
log2_sig_counts_symbol <- merge(log2_sig_counts, DEG_genenames_Th17, by="name")
rownames(log2_sig_counts_symbol) <- log2_sig_counts$name 
log2_sig_counts_symbol <- log2_sig_counts_symbol[, c(2:6,37)]
#write.table(log2_sig_counts_symbol, "Th17_log2_counts", sep="\t", quote=F)
```
[Download Th17 Results](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/Th17_results)

#### Significant differentially expressed genes - Disease vs. Healthy subset - expression patterns - Th17 cells
This plot shows the expression of the significant differentially expressed genes by sample. These genes do cluster by sample group; however, lab validation is still extremely important for any of the identified significant genes due to the high variability associated with the dataset. For the heatmap, the scale values compare the sample expression of a specific gene to the mean expression of the gene.
```{r heatmap_DE_genes_Th17}
get_heatmap_fn = function(summarydata_merged) {
        # return the pheatmap function with or without metadata
        if(ncol(Th17_meta) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=Th17_meta, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn = get_heatmap_fn(summarydata_merged)

log2_sig_counts <- log2_sig_counts[,1:5]
heatmap_fn(log2_sig_counts, clustering_method = "ward.D2", clustering_distance_cols = "correlation")

pca_matrix <- prcomp(t(log2_sig_counts))$x
df <- cbind(Th17_meta, pca_matrix[, c("PC1", "PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")
```
[Download Th17 rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/Th17_log2_counts)

### Genes of interest -  Disease vs. Healthy subset - Th17
Exploration of the counts and expression patterns for the genes of interest.

#### Genes of interest -- results table and counts - Th17
Among the genes of interest, none exhibited significant differential expression. 
```{r candidate_gene_table_Th17}
genes <- data.frame(baseMean = Th17[[1]], log2FoldChange = Th17[[2]], lfcSE = Th17[[3]], stat = Th17[[4]], pvalue = Th17[[5]], padj = Th17[[6]], row.names = row.names(Th17))
genes <- cbind(genes, name = row.names(genes))
genessymbol <- which(row.names(ann.genes1) %in% row.names(genes))
genessymbol <- ann.genes1[genessymbol,]
genessymbol <- cbind(genessymbol, name = row.names(genessymbol))
genes_genenames <- merge(genes, genessymbol, by="name")
row.names(genes_genenames) <- genes_genenames$name
genes_genenames <- genes_genenames[, c(2:7,32)]
CCL3 <- genes_genenames[which(genes_genenames$symbol=="CCL3"),]
CCL4 <- genes_genenames[which(genes_genenames$symbol=="CCL4"),]
CCL5 <- genes_genenames[which(genes_genenames$symbol=="CCL5"),]
CSF2 <- genes_genenames[which(genes_genenames$symbol=="CSF2"),]
CXCR3 <- genes_genenames[which(genes_genenames$symbol=="CXCR3"),]
EPSTI1 <- genes_genenames[which(genes_genenames$symbol=="EPSTI1"),]
GZMB <- genes_genenames[which(genes_genenames$symbol=="GZMB"),]
IFNG <- genes_genenames[which(genes_genenames$symbol=="IFNG"),]
IL23R <- genes_genenames[which(genes_genenames$symbol=="IL23R"),]
IL3 <- genes_genenames[which(genes_genenames$symbol=="IL3"),]
RGS2 <- genes_genenames[which(genes_genenames$symbol=="RGS2"),]
STAT1 <- genes_genenames[which(genes_genenames$symbol=="STAT1"),]
STAT3 <- genes_genenames[which(genes_genenames$symbol=="STAT3"),]
STAT5A <- genes_genenames[which(genes_genenames$symbol=="STAT5A"),]
TBX21 <- genes_genenames[which(genes_genenames$symbol=="TBX21"),]
marker_genes <- rbind(CCL3, CCL4, CCL5, CSF2, CXCR3, EPSTI1, GZMB, IFNG, IL23R, IL3, RGS2, STAT1, STAT3, STAT5A, TBX21)

marker_log2counts_Th17 <- as.data.frame(log2_counts[row.names(marker_genes),])
row.names(marker_log2counts_Th17) <- marker_genes$symbol
marker_genes <- marker_genes[,1:6]
knitr::kable(marker_genes)
#write.table(marker_log2counts_Th17, "Th17_marker_counts", sep="\t", quote=F)
```

#### Candidate genes -- expression patterns - Th17
The genes of interest did not cluster well by sample group for Th17 cells.

```{r candidate_gene_heatmap_Th17}

#Values relate to the row mean subtracted from the normalized count value for each sample value.
idx <- rownames(log2_counts) %in% rownames(marker_genes)
log2_marker_counts_Th17 <- log2_counts[idx,]
log2_marker_counts_Th17$name <- rownames(log2_marker_counts_Th17)
marker_genes$name <- rownames(marker_genes)
log2_marker_counts_Th17_symbol <- merge(log2_marker_counts_Th17, marker_genes, by="name")
row.names(log2_marker_counts_Th17_symbol) <- log2_marker_counts_Th17_symbol$symbol
log2_marker_counts_Th17_symbol <- log2_marker_counts_Th17_symbol[, c(2:6)]
heatmap_fn(log2_marker_counts_Th17_symbol, clustering_method = "ward.D2", clustering_distance_cols = "correlation")
```
[Download Marker Genes - rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/Th17_marker_counts)

### Functional Analysis - Th17 cells
Using the DE genes, a list of statistically enriched gene ontology (GO) terms was generated using the program [gprofileR](http://biit.cs.ut.ee/gprofiler/). From the two significant GO terms associated with the 124 DE genes, we can see that enrichment of gene ontology terms generates little information regarding affected pathways.
```{r functional_analysis_Th17}
gene_list <- DEG_genenames_Th17
gene_list <- cbind(gene_id = gene_list$symbol, gene_list)
gene_list <- gene_list[, c(1,7)]

#gprofileR
library(gProfileR)
gprofiler_results <- gprofiler(query = gene_list, organism = "hsapiens", ordered_query = F, 
                               exclude_iea = F, max_set_size = 0, correction_method = "fdr", 
                               hier_filtering = "none", domain_size = "annotated", custom_bg = "")

allterms <- gprofiler_results$term.id
GOs <- allterms[grep("GO:", allterms)]
pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
GO.pval <- cbind(gprofiler_results$term.id, gprofiler_results$p.value)
GO_genes_Th17 <- gprofiler_results[, c("term.id", "term.name", "p.value", "term.size", "overlap.size", "intersection")]
names(GO_genes_Th17) <- c("term.id", "term.name", "p.value", "term.size", "overlap.size", "assoc.gene.ids")
write.table(GO_genes_Th17, file ="GO_genes_Th17", quote = FALSE, sep = "\t", row.names=T)
```
[Download GO Terms](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/GO_genes_Th17)

## Differential Expression analysis - Disease vs. Healthy - Th1 cells
The following analysis performs differential expression testing for healthy controls versus patients in Th1 cells.
```{r de-setup_combined_Th1}
library(DESeq2)
library(vsn)
design = ~ pair + disease
condition = "disease"

# Differential expression design parameters

Th1_meta <- subset(combined_metadata, celltype == "Th1")
Th1_meta$disease <- factor(Th1_meta$disease)
Th1_meta$pair <- factor(Th1_meta$pair)
Th1_counts <- combined_counts_no_rrna[, row.names(Th1_meta)]

counts <- Th1_counts[rowSums(Th1_counts>0)>1,]
dds = DESeqDataSetFromMatrix(countData=counts,
                             colData=Th1_meta, design = design)
dds = DESeq(dds)

contrast2 <- list("diseasehealthy", "diseasepatient")
Th1 <- results(dds, contrast=contrast2)
#summary(Th1)
```

### Effect of variance stabilization
For RNA-Seq gene counts, the variance increases with the mean. To account for this variance, logarithmic transformation (log2) of normalized count values will ensure that these genes won't dominate the PCA plots. However, due to the noise associated with low count values, the general log2 transformation will worsen this noise, and low count genes will instead the PCA plots. Therefore, we need to use a transformation that will stabilize the variance across the mean for the gene counts. The plots below show the standard deviation of transformed counts using log2, rlog, and vsd transformations by rank(mean) (from top to bottom, respectively). The transformations are able to greatly reduce the standard deviation. The rlog transformation (middle) minimizes the variance best for the low and high counts, and stabilizes the variance nicely across the mean.

```{r deseq-diagnostics_Th1}
#For RNA-Seq raw counts, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and show the low count genes will dominate the PCA plots. Therfore, transform to stabilize variance across the mean using rlog. For high counts, gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

#Plotting standard deviation by rank(mean)
meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

### Similarity between samples
Since the rlog transformation best minimizes variance and standard deviation, we will use the rlog transformation to estimate variation between samples with Euclidean distance and PCA. 

### Dispersion estimates
The following plot shows the dispersion by mean of normalized counts. The dispersion looks good, since we expect the dispersion to decrease with increased mean of normalized counts. 

```{r dispersion-estimate_Th1}
plotDispEsts(dds)
```

### MA-plots
MA plots explore the mean expression level of the genes with the fold change, allowing visualization of the genes that are differentially expressed (red). Due to the large degree of variance within our sample groups, we are only able to detect genes with high expression and large fold changes between conditions.

```{r DESeq-MA_Th1}
#res <- results(dds, alpha=.05)
plotMA(Th1, ylim=c(-5,5))
```
### Volcano-plot
The following volcano plot visualizes the significant DE genes. Due to the variation in the data within the Th1 group, we are unable to detect any significant differential expression between healthy and diseased patients in Th1 cells for genes other than those with large fold changes. 
```{r DESeq-volcano_Th1}

df <- data.frame(Th1)
# ggplot(data=df, aes(x=log2FoldChange, y=-log10(padj))) +
#   scale_color_manual(values = c("grey", "purple")) +
#   xlim(c(-2,2)) +
#   ylim(c(0,1.3)) +
#   geom_point(alpha=0.75, pch=16) +
#   theme(legend.position = "none",
#         plot.title = element_text(size = rel(1.5)),
#         axis.title = element_text(size = rel(1.5)),
#         axis.text = element_text(size = rel(1.25))) +
#   xlab("log2 fold change") + ylab("-log10 p-value")

with(Th1, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-8,8), ylim=c(0,8)))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(Th1, padj<.05 ), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
with(subset(Th1, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="green"))
```

### Differentially expressed genes
Using an padj cut-off of 0.05 (p-values multiple test corrected using the BH method), 3 genes were identified as significant. 

#### Significant differentially expressed genes - Disease vs. Healthy subset - gene list - Th1 cells
The list of significant DE genes contains the gene expression differences in the patients  relative to the healthy controls. The row names in the table are the Ensembl gene ids, followed by the columns: the mean of the normalized counts for that gene for all samples (`baseMean`), log2 fold change (`log2FoldChange`), standard error (`lfcSE`), Wald statistic (`stat`), Wald test p-value (`pvalue`), BH adjusted p-values (`padj`), and the official gene symbol (`symbol`). 

The extremely short list of significant differentially expressed genes is likely due to the high variability seen within sample groups.

```{r DE_genes_list_Th1}
#Gene list with padj
resSig <- subset(Th1, padj < 0.05)
DEG <- data.frame(baseMean = resSig[[1]], log2FoldChange = resSig[[2]], lfcSE = resSig[[3]], stat = resSig[[4]], pvalue = resSig[[5]], padj = resSig[[6]], row.names = row.names(resSig))
DEG <- cbind(DEG, name = row.names(DEG))
DEGsymbol <- which(row.names(ann.genes1) %in% row.names(DEG))
DEGsymbol <- ann.genes1[DEGsymbol,]
DEGsymbol <- cbind(DEGsymbol, name = row.names(DEGsymbol))
DEGsymbol$name <- row.names(DEGsymbol) 
DEG_genenames_Th1 <- merge(DEG, DEGsymbol, by="name")

Th1_sig_norm_counts <- cbind(Th1_norm_counts, "name" = row.names(Th1_norm_counts))
Th1_sig_norm_counts <- merge(Th1_sig_norm_counts, DEG_genenames_Th1, by="name")
row.names(Th1_sig_norm_counts) <- Th1_sig_norm_counts$name
Th1_sig_norm_counts <- Th1_sig_norm_counts[, c(2:7,38)]

row.names(DEG_genenames_Th1) <- DEG_genenames_Th1$name
DEG_genenames_Th1 <- DEG_genenames_Th1[, c(2:7,32)]
DEG_genenames_Th1 <- DEG_genenames_Th1[order(DEG_genenames_Th1$padj),]
knitr::kable(DEG_genenames_Th1)
#write.table(Th1_sig_norm_counts, "Th1_norm_counts", sep="\t", quote=F)

#Normalized counts of genes
DEG_genenames_Th1 <- merge(DEG, DEGsymbol, by="name")
Th1_sig_counts <- cbind(Th1_counts, "name" = row.names(Th1_counts))
Th1_sig_counts <- merge(Th1_sig_counts, DEG_genenames_Th1, by="name")
row.names(Th1_sig_counts) <- Th1_sig_counts$name
Th1_sig_counts <- Th1_sig_counts[, c(2:7)]

log2_counts <- assay(rld[notAllZero,])
log2_counts <- as.data.frame(log2_counts) 
idx <- rownames(log2_counts) %in% rownames(Th1_sig_counts)
log2_sig_counts <- log2_counts[idx,]
log2_sig_counts$name <- rownames(log2_sig_counts)
log2_sig_counts_symbol <- merge(log2_sig_counts, DEG_genenames_Th1, by="name")
rownames(log2_sig_counts_symbol) <- log2_sig_counts$name 
log2_sig_counts_symbol <- log2_sig_counts_symbol[, c(2:7,38)]
#write.table(log2_sig_counts_symbol, "Th1_log2_counts", sep="\t", quote=F)
```
[Download Th1 Results](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/Th1_results)

#### Significant differentially expressed genes - Disease vs. Healthy subset - expression patterns -Th1 cells
This plot shows the expression of the significant differentially expressed genes by sample. These genes don't cluster by sample group, and there is a lot of variability in expression of these genes across the patient sample group. In addition, the differences in expression between patient and healthy samples does not appear great. Lab validation would be extremely important for any of the identified significant genes. The scale values compare the sample expression of a specific gene to the mean expression of the gene.
```{r heatmap_DE_genes_Th1}
get_heatmap_fn = function(summarydata_merged) {
        # return the pheatmap function with or without metadata
        if(ncol(Th1_meta) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=Th1_meta, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn = get_heatmap_fn(summarydata_merged)

log2_sig_counts <- log2_sig_counts[,1:6]
heatmap_fn(log2_sig_counts, clustering_method = "ward.D2", clustering_distance_cols = "correlation")

pca_matrix <- prcomp(t(log2_sig_counts))$x
df <- cbind(Th1_meta, pca_matrix[, c("PC1", "PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")
```
[Download Th1 rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/Th1_log2_counts)

### Genes of interest -  Disease vs. Healthy subset - Th1
Exploration of the counts and expression patterns for the genes of interest.

#### Genes of interest -- results table and counts - Th1
Among the genes of interest, none exhibited significant differential expression in Th1.

```{r candidate_gene_table_Th1}
genes <- data.frame(baseMean = Th1[[1]], log2FoldChange = Th1[[2]], lfcSE = Th1[[3]], stat = Th1[[4]], pvalue = Th1[[5]], padj = Th1[[6]], row.names = row.names(Th1))
genes <- cbind(genes, name = row.names(genes))
genessymbol <- which(row.names(ann.genes1) %in% row.names(genes))
genessymbol <- ann.genes1[genessymbol,]
genessymbol <- cbind(genessymbol, name = row.names(genessymbol))
genes_genenames <- merge(genes, genessymbol, by="name")
row.names(genes_genenames) <- genes_genenames$name
genes_genenames <- genes_genenames[, c(2:7,32)]
CCL3 <- genes_genenames[which(genes_genenames$symbol=="CCL3"),]
CCL4 <- genes_genenames[which(genes_genenames$symbol=="CCL4"),]
CCL5 <- genes_genenames[which(genes_genenames$symbol=="CCL5"),]
CSF2 <- genes_genenames[which(genes_genenames$symbol=="CSF2"),]
CXCR3 <- genes_genenames[which(genes_genenames$symbol=="CXCR3"),]
EPSTI1 <- genes_genenames[which(genes_genenames$symbol=="EPSTI1"),]
GZMB <- genes_genenames[which(genes_genenames$symbol=="GZMB"),]
IFNG <- genes_genenames[which(genes_genenames$symbol=="IFNG"),]
IL23R <- genes_genenames[which(genes_genenames$symbol=="IL23R"),]
IL3 <- genes_genenames[which(genes_genenames$symbol=="IL3"),]
RGS2 <- genes_genenames[which(genes_genenames$symbol=="RGS2"),]
STAT1 <- genes_genenames[which(genes_genenames$symbol=="STAT1"),]
STAT3 <- genes_genenames[which(genes_genenames$symbol=="STAT3"),]
STAT5A <- genes_genenames[which(genes_genenames$symbol=="STAT5A"),]
TBX21 <- genes_genenames[which(genes_genenames$symbol=="TBX21"),]
marker_genes <- rbind(CCL3, CCL4, CCL5, CSF2, CXCR3, EPSTI1, GZMB, IFNG, IL23R, IL3, RGS2, STAT1, STAT3, STAT5A, TBX21)

marker_log2counts_Th1 <- as.data.frame(log2_counts[row.names(marker_genes),])
row.names(marker_log2counts_Th1) <- marker_genes$symbol
marker_genes <- marker_genes[,1:7]
knitr::kable(marker_genes)
#write.table(marker_log2counts_Th1, "Th1_marker_counts", sep="\t", quote=F)
```

#### Genes of interest -- expression patterns - Th1
The genes of interest did not cluster well by sample group for Th1 cells.

```{r candidate_gene_heatmap_Th1}

#Values relate to the row mean subtracted from the normalized count value for each sample value.
idx <- rownames(log2_counts) %in% rownames(marker_genes)
log2_marker_counts_Th1 <- log2_counts[idx,]
log2_marker_counts_Th1$name <- rownames(log2_marker_counts_Th1)
marker_genes$name <- rownames(marker_genes)
log2_marker_counts_Th1_symbol <- merge(log2_marker_counts_Th1, marker_genes, by="name")
row.names(log2_marker_counts_Th1_symbol) <- log2_marker_counts_Th1_symbol$symbol
log2_marker_counts_Th1_symbol <- log2_marker_counts_Th1_symbol[, c(2:7)]
heatmap_fn(log2_marker_counts_Th1_symbol, clustering_method = "ward.D2", clustering_distance_cols = "correlation")
```
[Download Marker Genes - rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/Th1_marker_counts)

### Functional Analysis - Th1 cells
Using the DE genes, a list of statistically enriched gene ontology (GO) terms was generated using the program [gprofileR](http://biit.cs.ut.ee/gprofiler/). Many significant GO terms were associated with a single DE gene, ADRBK1. we can see that enrichment of gene ontology terms generates little information regarding affected pathways.

```{r functional_analysis_Th1}
gene_list <- DEG_genenames_Th1
gene_list <- cbind(gene_id = gene_list$symbol, gene_list)
gene_list <- gene_list[, c(1,7)]

#gprofileR
library(gProfileR)
gprofiler_results <- gprofiler(query = gene_list, organism = "hsapiens", ordered_query = F, 
                               exclude_iea = F, max_set_size = 0, correction_method = "fdr", 
                               hier_filtering = "none", domain_size = "annotated", custom_bg = "")

allterms <- gprofiler_results$term.id
GOs <- allterms[grep("GO:", allterms)]
pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
GO.pval <- cbind(gprofiler_results$term.id, gprofiler_results$p.value)
GO_genes_Th1 <- gprofiler_results[, c("term.id", "term.name", "p.value", "term.size", "overlap.size", "intersection")]
names(GO_genes_Th1) <- c("term.id", "term.name", "p.value", "term.size", "overlap.size", "assoc.gene.ids")
write.table(GO_genes_Th1, file ="GO_genes_Th1", quote = FALSE, sep = "\t", row.names=T)
```
[Download GO Terms](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/GO_genes_Th1)

## Differential Expression analysis - Disease vs. Healthy - DN cells
The following analysis performs differential expression testing for healthy controls versus patients in DN cells.
```{r de-setup_combined_DN}
library(DESeq2)
library(vsn)
design = ~ pair + disease
condition = "disease"

# Differential expression design parameters

DN_meta <- subset(combined_metadata, celltype == "DN")
DN_meta$disease <- factor(DN_meta$disease)
DN_meta$pair <- factor(DN_meta$pair)
DN_counts <- combined_counts_no_rrna[, row.names(DN_meta)]

counts <- DN_counts[rowSums(DN_counts>0)>1,]
dds = DESeqDataSetFromMatrix(countData=counts,
                             colData=DN_meta, design = design)
dds = DESeq(dds)

contrast2 <- list("diseasehealthy", "diseasepatient")
DN <- results(dds, contrast=contrast2)
#summary(DN)
```

### Effect of variance stabilization
For RNA-Seq gene counts, the variance increases with the mean. To account for this variance, logarithmic transformation (log2) of normalized count values will ensure that these genes won't dominate the PCA plots. However, due to the noise associated with low count values, the general log2 transformation will worsen this noise, and low count genes will instead the PCA plots. Therefore, we need to use a transformation that will stabilize the variance across the mean for the gene counts. The plots below show the standard deviation of transformed counts using log2, rlog, and vsd transformations by rank(mean) (from top to bottom, respectively). The transformations are able to reduce the standard deviation a bit.

```{r deseq-diagnostics_DN}
#For RNA-Seq raw counts, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and show the low count genes will dominate the PCA plots. Therfore, transform to stabilize variance across the mean using rlog. For high counts, gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

#Plotting standard deviation by rank(mean)
meanSdPlot(log2(counts(dds,normalized=TRUE)[notAllZero,] + 1),
           ylim = c(0,2.5))
meanSdPlot(assay(rld[notAllZero,]), ylim = c(0,2.5))
meanSdPlot(assay(vsd[notAllZero,]), ylim = c(0,2.5))
```

### Similarity between samples
Since the rlog transformation best minimizes variance and standard deviation, we will use the rlog transformation to estimate variation between samples with Euclidean distance and PCA. 

### Dispersion estimates
The following plot shows the dispersion by mean of normalized counts. The dispersion looks okay, since we expect the dispersion to decrease with increased mean of normalized counts. The dispersion is a bit scattered, but okay.

```{r dispersion-estimate_DN}
plotDispEsts(dds)
```

### MA-plots
MA plots explore the mean expression level of the genes with the fold change, allowing visualization of the genes that are differentially expressed (red). Due to the large degree of variance within our sample groups, we are only able to detect genes with large fold changes between conditions.

```{r DESeq-MA_DN}
#res <- results(dds, alpha=.05)
plotMA(DN, ylim=c(-5,5))
```

### Volcano-plot
The following volcano plot visualizes the significant DE genes. Due to the variation in the data within the DN group, we are unable to detect any significant differential expression between healthy and patients in DN cells for genes other than those with large fold changes. 
```{r DESeq-volcano_DN}

df <- data.frame(DN)
# ggplot(data=df, aes(x=log2FoldChange, y=-log10(padj))) +
#   scale_color_manual(values = c("grey", "purple")) +
#   xlim(c(-2,2)) +
#   ylim(c(0,1.3)) +
#   geom_point(alpha=0.75, pch=16) +
#   theme(legend.position = "none",
#         plot.title = element_text(size = rel(1.5)),
#         axis.title = element_text(size = rel(1.5)),
#         axis.text = element_text(size = rel(1.25))) +
#   xlab("log2 fold change") + ylab("-log10 p-value")

with(DN, plot(log2FoldChange, -log10(pvalue), pch=20, main="Volcano plot", xlim=c(-8,8), ylim=c(0,8)))

# Add colored points: red if padj<0.05, orange of log2FC>1, green if both)
with(subset(DN, padj<.05 ), points(log2FoldChange, -log10(pvalue), pch=20, col="red"))
with(subset(DN, padj<.05 & abs(log2FoldChange)>1), points(log2FoldChange, -log10(pvalue), pch=20, col="green"))
```

### Differentially expressed genes
Using an padj cut-off of 0.05 (p-values multiple test corrected using the BH method), 23 genes were identified as significant. 

#### Significant differentially expressed genes - Disease vs. Healthy subset - gene list - DN cells
The list of significant DE genes contains the gene expression differences in the patients  relative to the healthy controls. The row names in the table are the Ensembl gene ids, followed by the columns: the mean of the normalized counts for that gene for all samples (`baseMean`), log2 fold change (`log2FoldChange`), standard error (`lfcSE`), Wald statistic (`stat`), Wald test p-value (`pvalue`), BH adjusted p-values (`padj`), and the official gene symbol (`symbol`). 

The short list of significantly differentially expressed genes is likely due to the high variability seen within sample groups.

```{r DE_genes_list_DN}
#Gene list with padj
resSig <- subset(DN, padj < 0.05)
DEG <- data.frame(baseMean = resSig[[1]], log2FoldChange = resSig[[2]], lfcSE = resSig[[3]], stat = resSig[[4]], pvalue = resSig[[5]], padj = resSig[[6]], row.names = row.names(resSig))
DEG <- cbind(DEG, name = row.names(DEG))
DEGsymbol <- which(row.names(ann.genes1) %in% row.names(DEG))
DEGsymbol <- ann.genes1[DEGsymbol,]
DEGsymbol <- cbind(DEGsymbol, name = row.names(DEGsymbol))
DEG_genenames_DN <- merge(DEG, DEGsymbol, by="name")
row.names(DEG_genenames_DN) <- DEG_genenames_DN$name

DN_sig_norm_counts <- cbind(DN_norm_counts, "name" = row.names(DN_norm_counts))
DN_sig_norm_counts <- merge(DN_sig_norm_counts, DEG_genenames_DN, by="name")
row.names(DN_sig_norm_counts) <- DN_sig_norm_counts$name
DN_sig_norm_counts <- DN_sig_norm_counts[, c(2:7,38)]

DEG_genenames_DN <- DEG_genenames_DN[, c(2:7,32)]
DEG_genenames_DN <- DEG_genenames_DN[order(DEG_genenames_DN$padj),]
knitr::kable(DEG_genenames_DN)
#write.table(DEG_genenames_DN, "DN_results", sep="\t", quote=F)

#Normalized counts of genes
DEG_genenames_DN <- merge(DEG, DEGsymbol, by="name")
DN_sig_counts <- cbind(DN_counts, "name" = row.names(DN_counts))
DN_sig_counts <- merge(DN_sig_counts, DEG_genenames_DN, by="name")
row.names(DN_sig_counts) <- DN_sig_counts$name
DN_sig_counts <- DN_sig_counts[, c(2:7)]

log2_counts <- assay(rld[notAllZero,])
log2_counts <- as.data.frame(log2_counts) 
idx <- rownames(log2_counts) %in% rownames(DN_sig_counts)
log2_sig_counts <- log2_counts[idx,]
log2_sig_counts$name <- rownames(log2_sig_counts)
log2_sig_counts_symbol <- merge(log2_sig_counts, DEG_genenames_DN, by="name")
rownames(log2_sig_counts_symbol) <- log2_sig_counts$name 
log2_sig_counts_symbol <- log2_sig_counts_symbol[, c(2:7,38)]
#write.table(log2_sig_counts_symbol, "DN_log2_counts", sep="\t", quote=F)
```

[Download DN Results](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/DN_results)

#### Significant differentially expressed genes - Disease vs. Healthy subset - expression patterns -DN cells
This plot shows the expression of the significant differentially expressed genes by sample. These genes do  cluster by sample group; however, there is a lot of variability in expression of these genes across both sample groups. Lab validation would be extremely important for any of the identified significant genes. The scale values compare the sample expression of a specific gene to the mean expression of the gene.
```{r heatmap_DE_genes_DN}
get_heatmap_fn = function(summarydata_merged) {
        # return the pheatmap function with or without metadata
        if(ncol(DN_meta) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=DN_meta, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn = get_heatmap_fn(summarydata_merged)

log2_sig_counts <- log2_sig_counts[,1:6]
heatmap_fn(log2_sig_counts, clustering_method = "ward.D2", clustering_distance_cols = "correlation")

pca_matrix <- prcomp(t(log2_sig_counts))$x
df <- cbind(DN_meta, pca_matrix[, c("PC1", "PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = samplegroup), 
size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
0.3)) + ggtitle("PC1 vs PC2 :: normalized counts (no rrna)")
```
[Download DN rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/DN_log2_counts)

### Genes of Interest -  Disease vs. Healthy subset - DN
Exploration of the counts and expression patterns for the genes of interest.

#### Genes of Interest -- results table and counts - DN
Among the genes of interest, none exhibited significant differential expression.

```{r candidate_gene_table_DN}
genes <- data.frame(baseMean = DN[[1]], log2FoldChange = DN[[2]], lfcSE = DN[[3]], stat = DN[[4]], pvalue = DN[[5]], padj = DN[[6]], row.names = row.names(DN))
genes <- cbind(genes, name = row.names(genes))
genessymbol <- which(row.names(ann.genes1) %in% row.names(genes))
genessymbol <- ann.genes1[genessymbol,]
genessymbol <- cbind(genessymbol, name = row.names(genessymbol))
genes_genenames <- merge(genes, genessymbol, by="name")
row.names(genes_genenames) <- genes_genenames$name
genes_genenames <- genes_genenames[, c(2:7,32)]
CCL3 <- genes_genenames[which(genes_genenames$symbol=="CCL3"),]
CCL4 <- genes_genenames[which(genes_genenames$symbol=="CCL4"),]
CCL5 <- genes_genenames[which(genes_genenames$symbol=="CCL5"),]
CSF2 <- genes_genenames[which(genes_genenames$symbol=="CSF2"),]
CXCR3 <- genes_genenames[which(genes_genenames$symbol=="CXCR3"),]
EPSTI1 <- genes_genenames[which(genes_genenames$symbol=="EPSTI1"),]
GZMB <- genes_genenames[which(genes_genenames$symbol=="GZMB"),]
IFNG <- genes_genenames[which(genes_genenames$symbol=="IFNG"),]
IL23R <- genes_genenames[which(genes_genenames$symbol=="IL23R"),]
IL3 <- genes_genenames[which(genes_genenames$symbol=="IL3"),]
RGS2 <- genes_genenames[which(genes_genenames$symbol=="RGS2"),]
STAT1 <- genes_genenames[which(genes_genenames$symbol=="STAT1"),]
STAT3 <- genes_genenames[which(genes_genenames$symbol=="STAT3"),]
STAT5A <- genes_genenames[which(genes_genenames$symbol=="STAT5A"),]
TBX21 <- genes_genenames[which(genes_genenames$symbol=="TBX21"),]
marker_genes <- rbind(CCL3, CCL4, CCL5, CSF2, CXCR3, EPSTI1, GZMB, IFNG, IL23R, IL3, RGS2, STAT1, STAT3, STAT5A, TBX21)

marker_log2counts_DN <- as.data.frame(log2_counts[row.names(marker_genes),])
row.names(marker_log2counts_DN) <- marker_genes$symbol
marker_genes <- marker_genes[,1:7]
knitr::kable(marker_genes)
#write.table(marker_log2counts_DN, "DN_marker_counts", sep="\t", quote=F)
```

#### Genes of Interest -- expression patterns - DN
The genes of interest did not cluster well by sample group for DN cells.

```{r candidate_gene_heatmap_DN}

#Values relate to the row mean subtracted from the normalized count value for each sample value.
idx <- rownames(log2_counts) %in% rownames(marker_genes)
log2_marker_counts_DN <- log2_counts[idx,]
log2_marker_counts_DN$name <- rownames(log2_marker_counts_DN)
marker_genes$name <- rownames(marker_genes)
log2_marker_counts_DN_symbol <- merge(log2_marker_counts_DN, marker_genes, by="name")
row.names(log2_marker_counts_DN_symbol) <- log2_marker_counts_DN_symbol$symbol
log2_marker_counts_DN_symbol <- log2_marker_counts_DN_symbol[, c(2:7)]
heatmap_fn(log2_marker_counts_DN_symbol, clustering_method = "ward.D2", clustering_distance_cols = "correlation")
```
[Download Marker Genes - rlog normalized counts used to create PCA and heatmap figures](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/DN_marker_counts)

### Functional Analysis - DN cells
Using the DE genes, a list of statistically enriched gene ontology (GO) terms was generated using the program [gprofileR](http://biit.cs.ut.ee/gprofiler/). From the three significant GO terms associated with the 23 DE genes, we can see that enrichment of gene ontology terms generates little information regarding affected pathways.

```{r functional_analysis_DN}
gene_list <- DEG_genenames_DN
gene_list <- cbind(gene_id = gene_list$symbol, gene_list)
gene_list <- gene_list[, c(1,7)]

#gprofileR
library(gProfileR)
gprofiler_results <- gprofiler(query = gene_list, organism = "hsapiens", ordered_query = F, 
                               exclude_iea = F, max_set_size = 0, correction_method = "fdr", 
                               hier_filtering = "none", domain_size = "annotated", custom_bg = "")

allterms <- gprofiler_results$term.id
GOs <- allterms[grep("GO:", allterms)]
pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
GO.pval <- cbind(gprofiler_results$term.id, gprofiler_results$p.value)
GO_genes_DN <- gprofiler_results[, c("term.id", "term.name", "p.value", "term.size", "overlap.size", "intersection")]
names(GO_genes_DN) <- c("term.id", "term.name", "p.value", "term.size", "overlap.size", "assoc.gene.ids")
write.table(GO_genes_DN, file ="GO_genes_DN", quote = FALSE, sep = "\t", row.names=T)

#runRevigo(GOs, pvals)
```
[Download GO Terms](https://dl.dropboxusercontent.com/u/204381225/weiner/DE_analysis/GO_genes_DN)

```{r session_info, echo=FALSE}
sessionInfo()
```




