---    
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---
        
```{r setup0, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
                      message=FALSE, prompt=TRUE, comment='', fig.cap='', fig.height = 9, fig.width = 12, bootstrap.show.code=FALSE)
```

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Vafai RNA-seq - QC report"
clientname="Harveen Natarajan"
clientemail="harveen@broadinstitute.org"
labPI="Scott Vafai"
analystname="Mary Piper"
analystemail="piper@hsph.harvard.edu"
```
---
        
# Vafai RNA-Seq Quality Control Report
        
**Project description:**
RNA-Seq quality control report for the Vafai project investigating the neurodegenerative disorder, Leigh syndrome. Leigh syndrome includes the symptoms: reduced lifespan, retardation, lethargy, blindness, decreased motor skills, increased serum lactate, lesions in brain stem and part of cerebellum and optic valve. The mouse model for Leigh syndrome is Ndufs4-KO.  

**Experimental design for RNA-Seq:**
All samples were attained from the cerebellum/brainstem of male mice. The experimental groups are WT (n=8), Ndufs4-KO (n=8), and rapamycin-treated Ndufs4-KO (n=6). WT and KO mice are age-matched (24-26 days), but rapamycin-treated animals are not (94-140 days). 

**Client:** `r clientname`, `r labPI` group.  

**Analysts:** `r analystname` (`r analystemail`)

The most recent update of this html document occurred: `r date()`

---

# Quality control metrics

```{r qc-setup}
# Libraries

library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(RColorBrewer)
library(knitr)
library(knitrBootstrap)
library(tidyr)
library(reshape)
library(rmarkdown)
library(dplyr)
library(ggdendro)
library(grid)
library(reshape)
library(gridExtra)
library(Biobase)
library(scales)
library(DESeq2)
library(gProfileR)
library(DEGreport)
library(vsn)
library(ggrepel)
library(genefilter)

# Data files
project_summary = "~/Dropbox/HBC consults/2016_04_11_vafai/2016-04-19_vafai-rnaseq/project-summary.csv"
counts_file = "~/Dropbox/HBC consults/2016_04_11_vafai/2016-04-19_vafai-rnaseq/combined.counts"
ann_counts_file <- "~/Dropbox/HBC consults/2016_04_11_vafai/2016-04-19_vafai-rnaseq/annotated_combined.counts"

# Analysis variables
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
"#0072B2", "#D55E00", "#CC79A7")
summarydata = data.frame(read.table(project_summary, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata$Name = rownames(summarydata)
summarydata = summarydata[order(summarydata$Name),]
summary_no18 <- summarydata[-10,]

counts = read.table(counts_file, header=TRUE, row.names="id", check.names=FALSE)
counts = counts[, order(colnames(counts))]

colnames(counts) = gsub(".counts", "", colnames(counts))
# this is a list of all non user-supplied metadata columns that could appear

known_columns <- colnames(summarydata)

# Removing metadata column from known columns
known_columns <- known_columns[-21]
```


```{r heatmap-function}
## heatmap function to use later
get_heatmap_fn = function(summary_no18) {
    # return the pheatmap function with or without metadata
    metadata = summary_no18[, !colnames(summary_no18) %in% known_columns, drop=FALSE]
    if(ncol(metadata) == 0) {
       return(pheatmap)
    }
    else {
    # rownames(metadata) = summarydata$Name
    heatmap_fn = function(data, ...) {
        pheatmap(data, annotation=metadata, ...)
    }
    return(heatmap_fn)
}}
heatmap_fn = get_heatmap_fn(summary_no18)
```

## Total reads
The total number of reads for each sample varies greatly between samples (generally between 20 million to 100 million per sample), and sample RAPA18 has very few reads.

```{r total-reads}
ggplot(summarydata, aes(x=rownames(summarydata), y=Total.reads/1e6, fill=condition)) +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        geom_bar(stat = 'identity') +
        ylab("Total reads (million)") + xlab("")
```

## Mapped reads
The number of mapped reads corresponds to the number of total reads.

```{r mapped-plot}
ggplot(summarydata, aes(x=Name, y=Mapped/1e6, fill=condition)) +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    geom_bar(stat="identity") +
    ylab("Mapped reads (million)") + xlab("")
```

## Genomic mapping rate
The genomic mapping rate represents the percentage of reads mapping to the reference genome. Low mapping rates are indicative of sample contamination, poor sequencing quality or other artifacts.

The percent of input reads mapped is good (>75%) for nearly all of the samples. A bit lower for sample WT8, and RAPA18 is a clear outlier.

```{r mapping-rate-plot}
ggplot(summarydata, aes(x=Name, y=Mapping.Rate, fill=condition)) +
    geom_bar(stat="identity") +
    ylab("mapping rate") + xlab("") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90))
```

## Number of genes detected
The number of genes detected for each sample is good with ~20,000 genes detected for all samples except RAPA18.

```{r genes-detected-plot}
dd = data.frame(Name=names(counts), Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Name, y=Genes.Detected)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("")
```

## Gene detection saturation
Gene detection saturation was nearly reached for the majority of the samples, which means that most of the samples were able to detect the majority of expressed genes. While saturation of gene detection was at ~ 50 million reads mapped, reads that had ~ 25 million reads mapped still detected >20,000 genes.

```{r saturation-plot}
dd = data.frame(Mapped=summarydata$Mapped, Genes.Detected = colSums(counts > 0))
ggplot(dd, aes(x=Mapped/1e6, y=Genes.Detected)) +
    geom_point() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("genes detected") + xlab("reads mapped (million)")
```

## Exonic mapping rate
The exonic mapping rates are between 40 - 60%, which is not unusual for libraries created using Smart-Seq2 methods.

```{r exonic-mapping-plot}
ggplot(summarydata, aes(x=Name, y=Exonic.Rate, fill=condition)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
The samples exhibited very low rates of rRNA mapping, which is good.
```{r rRNA-rate-plot}
ggplot(summarydata, aes(x=Name, y=rRNA_rate, fill=condition)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("rRNA rate") + xlab("")
```

## Estimated fragment length of paired-end reads
Generally we expect the fragment length to be similar for all samples, which it is. We didn't include RAPA18 for visualization purposes (due to the poor mapping, RAPA18 had a much larger fragment length).

```{r fragment-length-plot}
summary_no18 <- summarydata[c(1:9, 11:22),]
ggplot(summary_no18, aes(x=Name, y=Average.insert.size, fill=condition)) +
    geom_bar(stat="identity") +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) +
    ylab("fragment length") + xlab("")
```

## Boxplot of log10 counts per gene
Generally, we expect similar count spreads for all genes between samples unless the library sizes are different. Since the library sizes varied substantially between samples, we see quite a bit of variation between samples.

```{r boxplot-raw}
melted = melt(counts)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Similar to the previous plot, only normalized for library size. The spread of the log10 TMM-normalized counts per gene data should be similar for every sample. The samples generally look good, except for RAPA18 is a definite outlier. 

Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

```{r boxplot-normalized}
y = DGEList(counts=counts)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
Generally, we expect similar count spreads for all genes between samples unless the library sizes are different. We see variation between the spreads for all samples, with RAPA18 as an obvious outlier. 

```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
    geom_density() +
    theme_bw(base_size=10) +
    theme(panel.grid.major = element_line(size = .5, color = "grey"),
          axis.text.x = element_text(angle=90)) + xlab("")
```

## Exploring similarity using principal component analysis (PCA) and heatmaps
Principal components analysis is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expression levels for genes and transforms it in principal component space, reducing each sample into one point. Thereby, we can separate samples by expression variation, and identify potential sample outliers. The PCA plot is a way to look at how samples are clustering. 

An intercorrelation heatmap is another way to look at how well samples cluster by plotting the correlation between the expression profiles of the samples.

These techniques (i.e. PCA and intercorrelation heatmaps) can identify outliers better when values have a similar dynamic range. We transformed the data to the log2 scale using rlog transformation (generated using [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)), which is better for datasets in which there are large differences in sequencing depth.

*Since the RAPA18 sample had few reads and is an outlier for all quality metrics, we will remove RAPA18 from further analysis in which it can skew the results.*

### PCA - all samples except RAPA18
In this PCA plot, the Control and Knockout groups do not cluster well, but the Rapamycin samples do. The sample WT8 appears to be an outlier. 

The lack of clustering between the WT and KO groups may suggest a problem for identifying differentially expressed genes. Potentially, there could be too great of variation within group compared to variation between groups to identify many differentially expressed genes. 


```{r sample_variation_pca_all_groups}
counts_no18 <- counts[, c(-10)]
normalized_counts_no18 <- normalized_counts[, c(-10)]

counts_no18 <- counts_no18[rowSums(counts_no18>0)>1,]
metadata <- summary_no18[, !colnames(summary_no18) %in% known_columns, drop=FALSE]


design = ~ condition
condition = "samplegroup"

dds = DESeqDataSetFromMatrix(countData=counts_no18,
    colData=metadata, design = design)
dds = DESeq(dds)


#For RNA-Seq raw counts, variance increases with the mean. Logarithmic transformation of normalized count values with a small pseudocount will account for large variations seen between the highest expressing genes so that these genes won't dominate the PCA plots. However, due to the strong noise among low count values due to Poisson, the general log2 transformation will amplify this noise, and show the low count genes will dominate the PCA plots. Therfore, transform to stabilize variance across the mean using rlog. For high counts, gives similar results as log2, but for low counts, values are shrunken towards the genes' average across samples.

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

## PCA - all groups
data <- plotPCA(rld, intgroup = "condition", returnData=TRUE)
percentVar <- round(100 * attr(data, "percentVar"))

ggplot(data, aes(PC1, PC2, color=condition)) + geom_point(size=4) + geom_text_repel(aes(label=names)) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance"))
```

### Heatmap - all samples except RAPA18
Similar to the PCA, the heatmap shows the Rapamycin group to cluster well, while the KO and WT samples do not. Also, WT8 appears to be an outlier.

```{r sample_variation_heatmap}

rlog_counts <- assay(rld[notAllZero,])

#Values relate to the row mean subtracted from the normalized count value for each sample value.
pheatmap(cor(rlog_counts), annotation=metadata, method="spearman")
```

### PCA - no RAPA
To explore how well the WT and KO samples cluster in the absence of the rapamycin group, we analyzed the expression profiles using only KO and WT samples.

In this PCA plot, the Control and Knockout groups still do not cluster well, and both WT8 and KO1 samples appear to be outliers.
```{r sample_variation_pca_no_rapa}

## PCA - no RAPA
metadata <- summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
idx <- which(metadata$condition != "RAPA")

metadata_norapa <- subset(metadata, condition != "RAPA")
counts_norapa <- counts[, idx]
counts_norapa <- counts_norapa[rowSums(counts_norapa>0)>1,]

design = ~ condition
condition = "samplegroup"

dds = DESeqDataSetFromMatrix(countData=counts_norapa,
    colData=metadata_norapa, design = design)
dds = DESeq(dds)

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

## PCA - no rapa
data <- plotPCA(rld, intgroup = "condition", returnData=TRUE)
# data <- plotPCA(rld, intgroup = "condition", returnData=TRUE, ntop=nrow(counts_norapa))
percentVar <- round(100 * attr(data, "percentVar"))

ggplot(data, aes(PC1, PC2, color=condition)) + geom_text_repel(aes(label=names)) +
        geom_point(size=4) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance"))
```

###Exploring similarity using heatmap - no rapa
Similar to the PCA, the samples in the heatmap still do not cluster very well. WT8 and KO1 also appear to be outliers.
```{r sample_variation_heatmap_no_rapa}

rlog_counts <- assay(rld[notAllZero,])

#Values relate to the row mean subtracted from the normalized count value for each sample value.
pheatmap(cor(rlog_counts), annotation=metadata_norapa, method="spearman")
```

### PCA - no outliers
The outlier samples WT8 and KO1 were removed from the groups to determine if the samples would cluster better without them. However, the wild type and knock-out samples still do not cluster by condition, and the higher variation within the sample groups than between sample groups may result in few differentially expressed genes. It is likely that removing these outliers will not greatly improve the differential expression results.

```{r sample_variation_pca_no_rapa_minus_outliers}

## PCA - no outliers
metadata <- summarydata[, !colnames(summarydata) %in% known_columns, drop=FALSE]
idx <- which(metadata$condition != "RAPA")

metadata_norapa <- subset(metadata, condition != "RAPA")
metadata_no_outliers <- metadata_norapa[c(-1,-16), ,drop=FALSE]

counts_no_outliers <- counts[, idx]
counts_no_outliers <- counts_no_outliers[, c(-1, -16)]
counts_no_outliers <- counts_no_outliers[rowSums(counts_no_outliers>0)>1,]

design = ~ condition
condition = "samplegroup"

dds = DESeqDataSetFromMatrix(countData=counts_no_outliers,
    colData=metadata_no_outliers, design = design)
dds = DESeq(dds)

par(mfrow=c(1,3))
notAllZero <- (rowSums(counts(dds))>0)
rld <- rlog(dds)
vsd <- varianceStabilizingTransformation(dds)
rlogMat <- assay(rld)
vstMat <- assay(vsd)

## PCA - no outliers
data <- plotPCA(rld, intgroup = "condition", returnData=TRUE)
percentVar <- round(100 * attr(data, "percentVar"))

ggplot(data, aes(PC1, PC2, color=condition)) + geom_text_repel(aes(label=names)) +
        geom_point(size=4) +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance"))

```

###Exploring similarity using heatmap - no outliers
The heatmap shows the samples appear cluster better by condition than in the PCA, but they still do not cluster very well. Similar to the PCA, the expression patterns for samples KO5 and KO15 tend to correlate more with the wild type samples, while WT10 appears to correlate more with the knock-out samples.
```{r sample_variation_heatmap_no_outliers}

rlog_counts <- assay(rld[notAllZero,])

#Values relate to the row mean subtracted from the normalized count value for each sample value.
pheatmap(cor(rlog_counts), annotation=metadata_no_outliers, method="spearman")
```

# Verification of Ndusf4 gene knock-outs
We generated the number of counts aligning to each exon for Ndusf4 gene. All KO samples were missing reads for exon 2 (denoted as ENSMUSG00000021764:004 in the table), but exhibited expression for all other exons of Ndufs4. All WT samples expressed all exons of the Ndufs4 gene.

Below are the read counts aligning to each Ndufs4 exon for each sample:
```{r ndufs4_counts}
ndufs4_counts <- data.frame(read.table("../vafai_ndufs4_expression.txt", header=TRUE, sep="\t"), row.names="id", check.rows=FALSE)
ndufs4_counts
```

We also visualized the read alignments using IGV. Exon 2 is shown for all samples, while the KO and RAPA samples have exon 3 images as well, to show these samples have normal expression of other exons:

[KO samples - exon 2](https://dl.dropboxusercontent.com/u/204381225/vafai/ndufs4_img/vafai_KO_ndufs4_exon2.png)

[KO samples - exon 3](https://dl.dropboxusercontent.com/u/204381225/vafai/ndufs4_img/vafai_KO_ndufs4_exon3.png)

[WT samples - exon 2](https://dl.dropboxusercontent.com/u/204381225/vafai/ndufs4_img/vafai_WT_exon2.png)

[RAPA samples - exon 2](https://dl.dropboxusercontent.com/u/204381225/vafai/ndufs4_img/vafai_RAPA_ndufs4_exon2.png)

[RAPA samples - exon 3](https://dl.dropboxusercontent.com/u/204381225/vafai/ndufs4_img/vafai_RAPA_ndufs4_exon3.png)


***Based on the QC analysis, we will explore differential expression analysis with and without the K01 and WT8 outliers. However, there is little difference between the expression profiles for WT and KO mice, and the variation within group may be greater than the variation between groups to identify many differentially expressed genes.*** 

```{r session_info, results="hide"}
sessionInfo()
```