---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

        
```{r setup0, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
                      cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE, message=FALSE, prompt=TRUE, comment='', fig.cap='', fig.height = 9, fig.width = 12, bootstrap.show.code=FALSE)
```

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Macklis RNA-seq - run1"
clientname="Hari Padmanabhan"
clientemail="hari_padmanabhan@harvard.edu"
labPI="Jeffrey Macklis"
analystname="Mary Piper and Lorena Pantano"
analystemail="piper@hsph.harvard.edu and lpantano@hsph.harvard.edu"
```
---
        
# Macklis Quality Control Report
        
Quality control report for RNA-Seq project on Sox6 heterozygous (Ctrl) and knockout (KO) oligodendrocytes isolated from the cortex of age-matched neonatal mouse brains.

Client: `r clientname`, `r labPI` group.  

Analysts: `r analystname` (`r analystemail`)

The most recent update of this html document occurred: `r date()`

---
        
# Overview
        
```{r qc-setup}
library(ggplot2)
library(reshape)
library(gplots)
library(edgeR)
library(CHBUtils)
library(pheatmap)
library(RColorBrewer)
library(knitr)
library(tidyr)
library(reshape)
library(rmarkdown)
library(dplyr)
library(ggdendro)
library(grid)
library(reshape)
library(gridExtra)
library(Biobase)
library(scales)
library(DESeq2)
project_summary3 = "~/Dropbox/HBC\ consults/2015_11_24_macklis/run3/2015-12-02_macklis-rnaseq3/project-summary3.csv"
counts3_file = "~/Dropbox/HBC\ consults/2015_11_24_macklis/run3/2015-12-02_macklis-rnaseq3/combined.counts"
annotat3_counts_file <- "~/Dropbox/HBC\ consults/2015_11_24_macklis/run3/2015-12-02_macklis-rnaseq3/annotated_combined.counts"
project_summary2 = "~/Dropbox/HBC\ consults/2015_11_24_macklis/run2/2015-12-01_macklis-rnaseq/project-summary-run2.csv"
counts2_file = "~/Dropbox/HBC\ consults/2015_11_24_macklis/run2/2015-12-01_macklis-rnaseq/combined.counts"
project_summary1 = "~/Dropbox/HBC\ consults/2015_11_24_macklis/run1/2015-11-26_macklis-rnaseq_unst_run1/project-summary_unst_run1.csv"
counts1_file = "~/Dropbox/HBC\ consults/2015_11_24_macklis/run1/2015-11-26_macklis-rnaseq_unst_run1/combined.counts"
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442",
               "#0072B2", "#D55E00", "#CC79A7")
#Run 3
summarydata3 = data.frame(read.table(project_summary3, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata3$Name = rownames(summarydata3)
summarydata3 = summarydata3[order(summarydata3$Name),]
counts3 = read.table(counts3_file, header=TRUE, row.names="id", check.names=FALSE)
counts3 = counts3[, order(colnames(counts3))]
colnames(counts3) = gsub(".counts", "", colnames(counts3))
total_counts3 <- colSums(counts3, na.rm = FALSE, dims = 1)
annot_counts3 <- read.table(annotat3_counts_file, header=TRUE, row.names="id", check.names=FALSE)
annot_counts3 = annot_counts3[, order(colnames(annot_counts3))]

## this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
                  "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
                  "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
                  "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
                  "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
                  "complexity", "X5.3.bias")
metadata3<- summarydata3[,c(25,26,27)]
metadata3<-metadata3[,c(1,3)]

#Run 2
summarydata2 = data.frame(read.table(project_summary2, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata2$Name = rownames(summarydata2)
summarydata2 = summarydata2[order(summarydata2$Name),]
counts2 = read.table(counts2_file, header=TRUE, row.names="id", check.names=FALSE)
counts2 = counts2[, order(colnames(counts2))]
colnames(counts2) = gsub(".counts", "", colnames(counts2))
total_counts2 <- colSums(counts2, na.rm = FALSE, dims = 1)

## this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
                  "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
                  "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
                  "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
                  "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
                  "complexity", "X5.3.bias")
metadata2<- summarydata2[,c(25,26,27)]
metadata2<-metadata2[,c(1,3)]

#Run 1
summarydata1 = data.frame(read.table(project_summary1, header=TRUE, sep=","), row.names="Name", check.rows=FALSE)
summarydata1$Name = rownames(summarydata1)
summarydata1 = summarydata1[order(summarydata1$Name),]
counts1 = read.table(counts1_file, header=TRUE, row.names="id", check.names=FALSE)
counts1 = counts1[, order(colnames(counts1))]
colnames(counts1) = gsub(".counts", "", colnames(counts1))
total_counts1 <- colSums(counts1, na.rm = FALSE, dims = 1)

## this is a list of all non user-supplied metadata columns that could appear
known_columns = c("Name", "X.GC", "Exonic.Rate", "Sequences.flagged.as.poor.quality",
                  "rRNA_rate", "Fragment.Length.Mean", "Intronic.Rate", "Intergenic.Rate",
                  "Mapping.Rate", "Quality.format", "Duplication.Rate.of.Mapped", "Mapped",
                  "rRNA", "Sequence.length", "Transcripts.Detected", "Mean.Per.Base.Cov.",
                  "Genes.Detected", "Unique.Starts.Per.Read", "unique_starts_per_read",
                  "complexity", "X5.3.bias")
metadata1<- summarydata1[,c(25,26,27)]
metadata1<-metadata1[,c(1,3)]
#Merge
summarydata_merge <- rbind(summarydata1, summarydata2, summarydata3)
counts_merge <- cbind(counts1, counts2, counts3)
metadata_merge <- rbind(metadata1, metadata2, metadata3)
```

```{r heatmap-function}
get_heatmap_fn = function(summarydata_merge) {
        # return the pheatmap function with or without metadata
        if(ncol(metadata_merge) == 0) {
                return(pheatmap)
        }
        else {
                # rownames(metadata) = summarydata$Name
                heatmap_fn = function(data, ...) {
                        pheatmap(data, annotation=metadata_merge, ...)
                }
                return(heatmap_fn)
        }}
heatmap_fn = get_heatmap_fn(summarydata_merge)
```

# Quality control metrics

The samples for each of the individual sequencing runs are shown in the following plots, using the nomenclature "samplename_run#". For example, "BC1_2" refers to sample "BC1", run 2.

We use the following quality control metrics to assess the quality of the samples and identify outliers and to determine whether we see any strong batch effects. If we do not see any strong batch effects, then we will combine all runs for each sample for the differential expression analysis.

## Total reads
The total reads are the number of input reads for each sample. The number of reads for each sample is similar between samples for each run, which is good. We start with ~5 million reads for runs 1 and 2, and >20 million reads for run3. 20 million reads is low for general RNA-Seq analyses, but quite good for low input RNA-Seq.

```{r table, results='asis'}
path_results = "~/Dropbox/HBC\ consults/2015_11_24_macklis/"
qc = read.table(file.path(path_results, "metrics_merge.tsv"),
                header=T, sep="\t", check.names=F,
                colClasses=list("sample"="character"))
rownames(qc) = qc$sample
qc$Mapped_reads_pct = as.numeric(gsub("%", "", qc$Mapped_reads_pct))
qc$Duplicates_pct = as.numeric(gsub("%", "", qc$Duplicates_pct))
qc$samplegroup <- metadata_merge$samplegroup

metrics = c("sample", "Total_reads" ,"Mapped_reads_pct", "Duplicates_pct",
            "offtarget", "%GC", "Sequence_length", "Median_insert_size")
```

```{r total-reads}
ggplot(qc, aes(x=sample, y=Total_reads/1e6, fill=samplegroup)) +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        geom_bar(stat = 'identity') +
        ylab("Million reads")
```

## Mapped reads
Most of the reads mapped with ~5 million reads mapping for runs 1 and 2, and >20 million for run3.

```{r mapped-plot}
ggplot(summarydata_merge, aes(x=Name, y=Mapped, fill=samplegroup)) +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        geom_bar(stat="identity") +
        ylab("mapped reads") + xlab("")
```

## Genomic mapping rate
The mapping rate is very good, with ~90% of reads mapping to the genome. Also, good is that there are similar mapping rates for all samples, and no obvious outliers.

```{r mapping-rate-plot}
ggplot(summarydata_merge, aes(x=Name, y=Mapping.Rate, fill=samplegroup)) +
        geom_bar(stat="identity") +
        ylab("mapping rate") + xlab("") +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90))
```

## Number of genes detected
The number of genes detected for each sample is good for the number of input reads. Over 15,000 genes are detected for runs 1 and 2, and ~18,000 genes detected for run3. The greater number of genes detected for run3 is to be expected based on the greater depth of sequencing for this run.

```{r genes-detected-plot}
dd <- data.frame(Name=names(counts_merge), Genes.Detected = colSums(counts_merge > 0))
dd$samplegroup <- metadata_merge$samplegroup
ggplot(dd, aes(x=Name, y=Genes.Detected, fill=samplegroup)) +
        geom_bar(stat="identity") +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        ylab("genes detected") + xlab("")
```

## Gene detection saturation
Gene detection saturation was not reached for samples in this experiment. The greater number of reads mapped, the greater the number of genes detected.

```{r saturation-plot}
dd = data.frame(Mapped=summarydata_merge$Mapped, Genes.Detected = colSums(counts_merge > 0))
ggplot(dd, aes(x=Mapped, y=Genes.Detected)) +
        geom_point() +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        ylab("genes detected") + xlab("reads mapped")
```

## Exonic mapping rate
While generally exonic mapping rates for RNA-Seq analyses should be ~70%, the method used for cDNA generation and library preparation from low input RNA is expected to yield much lower exonic mapping rates ~40%. Therefore, the ~50% exonic mapping rates are quite good. Slightly higher exonic mapping rates for KO samples, but still quite similar to Ctrl samples.

```{r exonic-mapping-plot}
ggplot(summarydata_merge, aes(x=Name, y=Exonic.Rate, fill=samplegroup)) +
        geom_bar(stat="identity") +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        ylab("exonic mapping rate") + xlab("")
```

## rRNA mapping rate
Very low rates of rRNA mapping, which is good.

```{r rRNA-rate-plot}
ggplot(summarydata_merge, aes(x=Name, y=rRNA_rate, fill=samplegroup)) +
        geom_bar(stat="identity") +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        ylab("rRNA rate") + xlab("")
```

## Estimated fragment length of paired-end reads
Generally we expect the fragment length to be similar for all samples. We are looking into why we see such drastically different fragment lengths between samples, and may be due to the method used to calculate the fragment length. Since all other QC metrics look good, we are not going to worry too much about it.

```{r fragment-length-plot}
ggplot(summarydata_merge, aes(x=Name, y=Fragment.Length.Mean, fill=samplegroup)) +
        geom_bar(stat="identity") +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) +
        ylab("fragment length") + xlab("")
```

## Boxplot of log10 counts per gene
Generally, we expect similar count spreads for all genes between samples unless the library sizes are different. Since the run3 library was much bigger, we see an increased number of counts on average and a larger spread, which is to be expected.

```{r boxplot-raw}
melted = melt(counts_merge)
colnames(melted) = c("sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) + xlab("")
```

## Boxplot of log10 TMM-normalized counts per gene
Trimmed mean of M-values (TMM) normalization is described
[here](http://genomebiology.com/2010/11/3/R25)

Robinson, M. D., & Oshlack, A. (2010). A scaling normalization method for differential expression analysis of RNA-seq data. Genome Biology, 11(3). doi:10.1186/gb-2010-11-3-r25

Similar to the previous plot, only normalized for library size. Now distributions of counts per gene appear similar between samples. The larger spread for the run3 samples is likely due to the ability to detect some additional low expressing genes.

```{r boxplot-normalized}
y = DGEList(counts=counts_merge)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
melted = melt(normalized_counts)
colnames(melted) = c("gene", "sample", "count")
melted$sample = factor(melted$sample)
melted = melted[order(melted$sample),]
melted$count = log(melted$count)
ggplot(melted, aes(x=sample, y=count)) + geom_boxplot() +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) + xlab("")
```

## Density of log10 TMM-normalized counts
Generally, the density of log10 TMM-normalized counts should be similar for all samples based on the genes detected. Generally, each run yields similar densities between samples, which is good. No obvious outliers. Again, run3 is able to detect more low expressing genes, so there is a slight shift in the lower part of the curve for run3.

```{r density-normalized}
ggplot(melted, aes(x=count, group=sample)) +
        geom_density() +
        theme_bw(base_size=10) +
        theme(panel.grid.major = element_line(size = .5, color = "grey"),
              axis.text.x = element_text(angle=90)) + xlab("")
```

## Correlation (Pearson) heatmap of TMM-normalized counts
The Ctrl and KO samples cluster nicely.

```{r pearson-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="pearson"))
```

## Correlation (Spearman) heatmap of TMM-normalized counts
The Spearman correlation heatmap is more resistant to outliers than the Pearson heatmap (above), and again, the groups cluster well.

```{r spearman-heatmap-normalized}
heatmap_fn(cor(normalized_counts, method="spearman"))
```

## PCA plot of TMM-normalized counts
Principal components analysis is a multivariate technique that allows us to summarize the systematic patterns of variations in the data. PCA takes the expression levels for genes and transforms it in principal component space, reducing each sample into one point. Thereby, we can separate samples by expression variation, and identify potential sample outliers. The PCA plot is another way to look at how samples are clustering. 

In this PCA plot the different sequencing runs for the same samples cluster well, so we should be able to combine the counts for each run together for the differential expression analysis. The Control and Knockout groups do not cluster together, so it may suggest a problem for identifying differentially expressed genes. Potentially, there could be too great of variation within group compared to variation between groups to identify many differentially expressed genes.
```{r pca}
## MDS plot of TMM-normalized counts
# mds(normalized_counts, k=length(colnames(normalized_counts)) - 1)
metadata = subset(metadata_merge, select= c(samplegroup,Name))

pca_matrix <- prcomp(t(normalized_counts))$x
df <- cbind(metadata, pca_matrix[, c("PC1", "PC2")])

ggplot(df, aes(PC1, PC2, color = samplegroup)) + geom_text(aes(PC1, PC2, label = Name), 
    size = 5, hjust = 0.1, vjust = 0.1) + scale_x_continuous(expand = c(0.3, 
    0.3)) + ggtitle("PC1 vs PC2 :: normalized counts")

## Heatmap of top 30 most expressed genes
#select = order(rowMeans(counts_merge),decreasing=TRUE)[1:30]
#heatmap_fn(counts_merge[select,])
```

## Heatmap by concordance correlation coefficient
http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1004075

```{r propcor-heatmap}
propcor = function(x, y) {
        x = log(x + 0.1)
        y = log(y + 0.1)
        num = 2 * cov(x, y)
        denom = var(x) + var(y)
        return(num/denom)}

do_propcor = function(x) {
        mat = list()
        for(i in seq_len(ncol(x))) {
                for(j in seq_len(ncol(x))) {
                        x2 = x[, i]
                        y2 = x[, j]
                        mat = c(mat, propcor(x2, y2)) } }
        mat = unlist(mat)
        mat = matrix(mat, ncol(x), ncol(x))
        colnames(mat) = colnames(x)
        rownames(mat) = colnames(x)
        return(mat)}

heatmap_fn(do_propcor(normalized_counts))
```

#Examination of gene counts for specific genes: 
To determine whether all cells came from animals of the same sex, counts for Xist gene were examined. The Xist gene is only expressed for X-chromosome inactivation, so only females should show expression of this gene. BC1, BC2, and BC3 are male and BC4 is female. We may need to account for sex during the differential expression analysis.

As a positive control for the analysis, the counts for the gene used to FACS sort the oligodendrocyte cells, Cspg4, was examined. As expected, high expression of the gene is observed in all samples.

Finally, we observed the counts for genes of interest. Only a single gene in the set appears to have count differences between Ctrl and KO samples, CD9. However, these are not normalized for library size, so differential expression analysis is needed to determine if genes are differentially expressed.

```{r imp_genes}
annot_counts3_cut <- annot_counts3
annot_counts3_cut$Name <- row.names(annot_counts3_cut)
annot_counts3_cut <- annot_counts3_cut[,5:6]
Cspg4 <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000032911")
Sox6 <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000051910")
Xist <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000086503")
Tcf7l2 <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000024985")
Nkx2.2 <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000027434")
Casr <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000051980")
Parvb <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000022438")
APC <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000005871")
Fyn <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000019843")
CD9 <- subset(counts_merge, row.names(counts_merge) == "ENSMUSG00000030342")

important_genes <- data.frame()
important_genes <- rbind(important_genes, Cspg4, Sox6, Xist, Tcf7l2, Nkx2.2, Casr, Parvb, APC, Fyn, CD9)
important_genes$Name <- row.names(important_genes)
important_genes <- merge(important_genes, annot_counts3_cut, by= "Name")
knitr::kable(important_genes)
```

## KO Samples in Sox6
While expression of exon 4 of Sox6 is evident for the Ctrl heterozygous samples, no expression of exon 4 was visualized for the KO samples, as expected.

![Sox6 expression - exon4](Sox6_exon4.png)

However, all samples exhibit expression of exon 3 and exon 5. Since 2 exons were supposed to be deleted, let us know if this is a problem.

![Sox6 expression - exon3](Sox6_exon3.png)