---
output:
  knitrBootstrap::bootstrap_document:
    theme: readable
    highlight: zenburn
    theme.chooser: TRUE
    highlight.chooser: TRUE
  html_document:
    toc: true
    highlight: zenburn
---

```{r setup0, echo=FALSE}
knitr::opts_chunk$set(tidy=TRUE, highlight=TRUE, dev="png",
               cache=FALSE, highlight=TRUE, autodep=TRUE, warning=FALSE, error=FALSE,
               message=FALSE, prompt=TRUE, comment='', fig.cap='', fig.height = 9, fig.width = 12, bootstrap.show.code=FALSE)
```

```{r projsetup, echo=FALSE, warning=FALSE, message=FALSE}
project="Breault RNA-seq"
clientname="Emanuele Pignatti"
clientemail=""
labPI="David Breault"
analystname="Mary Piper and Meeta Mistry"
analystemail="piper@hsph.harvard.edu and mmistry@hsph.harvard.edu"
```
---

# Breault Differential Gene Expression Report

Differential gene expression report for RNA-Seq on adrenal glands from BMAL-KO mice exposed to normal lighting conditions (12hr light and 12hr darkness) and stressful lighting conditions (3hr light and 3hr darkness).

Client: `r clientname`, `r labPI` group.  

Analysts: `r analystname` (`r analystemail`)

The most recent update of this html document occurred: `r date()`

---
# Setup
```{r de set-up, echo=FALSE}
##Bioconductor and CRAN libraries used
library(ggplot2)
library(gplots)
library(RColorBrewer)
library(CHBUtils)
library(ggdendro)
library(grid)
library(reshape)
library(gridExtra)
library(edgeR)
library(Biobase)
library(gProfileR)
library(pheatmap)
library(scales)
source('~/Dropbox/HBC consults/2015_09_08_breault/revigo.R')

#Set variables
baseDir <- getwd()
dataDir <- paste(baseDir, "/data", sep="")
resultsDir <- paste(baseDir, "/results", sep="")
metaDir <- paste(baseDir, "/meta", sep="")

#Load data and metadata
data <- read.delim(file.path(dataDir, 'combined.counts'), header=T, sep="\t", row.names=1, as.is=T)
ann.genes <- read.delim(file.path(dataDir, 'annotated_combined.counts'), header=T, sep="\t", row.names=1, as.is=T)
colnames(data) <- c("S4", "S5", "S6", "S7", "S8", "S9", "S10")
meta <- read.delim(file.path(metaDir, 'breault-rnaseq.csv'), header=T, sep=",", row.names=1)
rownames(meta) <- c("S4", "S5", "S6", "S7", "S8", "S9", "S10")

#Calculate normalized count values
y = DGEList(counts=data)
y = calcNormFactors(y)
normalized_counts = cpm(y, normalized.lib.sizes=TRUE)
ann.norm.counts <- merge(normalized_counts, ann.genes, by=0)
row.names(ann.norm.counts) <- ann.norm.counts$Row.names
ann.norm.counts <- ann.norm.counts[,c(2:8, 16)]
#write.table(ann.norm.counts, file ="annotated_normalized_counts", quote = FALSE, sep = "\t", row.names=T)
```

# Differential Expression Analysis
Differential gene expression analysis of count data was performed using the Bioconductor R package, [EdgeR](http://bioconductor.org/packages/release/bioc/html/edgeR.html). The count data was fit to a negative binomial model and dispersion estimates were generated using the quantile-adjusted conditional maximum likelihood (qCML) method for single factor comparisons. Since differential expression testing was performed for a single factor (lighting conditions), the exact test for the negative binomial distribution was used to test for differential gene expression.

```{r de analysis}
#Filter out genes with zero counts
rowSums <- apply(data, 1, sum)
keep <- which(rowSums > 0)
data_filtered <- data[keep,]

#Create DGE object with data and metadata
dge <- DGEList(counts=data_filtered, group=meta$samplegroup)

#Normalize counts by total count
dge <- calcNormFactors(dge)
#normalized_counts2 = cpm(dge, normalized.lib.sizes=TRUE)
#dge$samples

#Estimate dispersion parameter for two sample comparison
dge <- estimateCommonDisp(dge, verbose= TRUE)
dge <- estimateTagwiseDisp(dge, verbose = TRUE)

#Model fitting
et <- exactTest(dge)
full_gene_stats <- topTags(et, n=23935)
#write.table(full_gene_stats, file ="All_genes_stats", quote = FALSE, sep = "\t", row.names=T)
```
[Download list of differential expression analysis results for all genes](https://dl.dropboxusercontent.com/u/204381225/breault/results/all_genes_stats)

[Download list of normalized counts for all genes](https://dl.dropboxusercontent.com/u/204381225/breault/results/annotated_normalized_counts)

# Significant Differentially Expressed (DE) Genes
Using an FDR cut-off of 0.05 (p-values multiple test corrected using the BH method), only 13 genes were identified as significant. The expression changes listed below for all analyses describe differences in the stressed lighting T7 mice relative to the control T24 mice. 

In the table below, the row values indicate whether the gene is up- or down-regulated (1, and -1 respectively). For example, the first row indicates that the expression levels of five genes were significantly down-regulated in T7 mice relative to T24 mice.

```{r significant genes summary}
summary(de <- decideTestsDGE(et, p=0.05))
```

### List of significant DE genes (FDR = 0.05)
The list of significant DE genes contains the gene expression differences in the stressed lighting T7 mice relative to the control T24 mice. The row names in the table are the Ensembl gene ids, followed by the columns: log2 fold change (`logFC`), average log2 counts per million, normalized for library sizes (`logCPM`), p-value (`PValue`), padj-value (`FDR`), differential expression indicator (`de`: whether the gene is up- or down-regulated) and the official gene symbol (`Symbol`). 

For example, in the first row of the table for gene id ENSMUSG00000042359 (official gene symbol = Osbp16), the T7 mice had an expression level log2 fold change of 0.791 relative to the T24 mice, with an average log CPM of 6.268, and the gene was up-regulated relative to T24 mice. The small list of significant DE genes is likely due to the high variability of expression within groups identified in the QC report.

```{r significant genes}
siget <- et[[1]]
siget <- cbind(siget, de)
sigde <- subset(siget, de != 0)
top13 <- topTags(et, n=13)
top13 <- top13[[1]]
top13$PValue <- format(top13$PValue, scientific = TRUE)
top13$FDR <- format(top13$FDR, scientific = TRUE)
sig.ann.genes <- ann.genes[row.names(final),]
sigde <- cbind(sigde, Symbol = sig.ann.genes$symbol)
finalsig <- merge(top13, sigde, by=0)
finalsig <- finalsig[,c(1:5,9,10)]
finalsig <- transform(finalsig, CPM = 2^logCPM.x)
names(finalsig) <- c("Ensembl_IDs", "logFC", "logCPM", "PValue", "FDR", "de", "Symbol", "CPM")
finalsig <- finalsig[, c(1:3, 8, 4:7)]
knitr::kable(finalsig)
#write.table(finalsig, file ="sig_de_genes", quote = FALSE, sep = "\t", row.names=T)
```
[Download list of significant DE genes](https://dl.dropboxusercontent.com/u/204381225/breault/results/sig_de_genes)

### Volcano plot of significant DE genes (FDR = 0.05)
The following volcano plot visualizes the significant DE genes (threshold FDR = 0.05) in purple relative to the genes not significantly differentially expressed (in grey). This plot indicates that many of the genes with large fold changes (logFC > 1) between conditions are not significant at our designated threshold. 

```{r volcano plot - sig genes}

df.siget <- data.frame(siget)
df.siget$threshold <- as.logical(df.siget$de)

ggplot(data = df.siget, aes(x = siget$logFC, y = -log10(PValue), 
    colour = threshold)) + scale_color_manual(values = c("grey", "purple")) + 
    geom_point(alpha = 0.75, pch = 16) + xlim(c(-10, 10)) + theme(legend.position = "none", 
    plot.title = element_text(size = rel(1.25)), axis.title = element_text(size = rel(1)), 
    axis.text = element_text(size = rel(0.75))) + ggtitle("Volcano plot of significantly DE genes") + 
    xlab("log2 fold change") + ylab("-log10 p-value") 
```

### Smear plot of significant DE genes (FDR = 0.05)
The smear plot visualizes the significant DE genes (threshold FDR = 0.05) in red relative to the genes not differentially expressed at that threshold (in black). The blue lines indicate a 2-fold change in expression between conditions. This plot indicates that many of the genes associated with large fold changes between conditions were not identified as significant due to low expression levels (small count values).

```{r smear plot - sig genes}
detags <- rownames(dge)[as.logical(de)]
plotSmear(et, de.tags=detags)
abline(h=c(-1, 1), col="blue")
```

# Top 50 Most Differentially Expressed Genes
Since only a few genes were identified as significant using the FDR cut-off value of 0.05, the top 50 differentially expressed genes were used for further analysis (although many will be outside of the significance threshold value of FDR = 0.05). The table and plots below characterize the variation, fold change values, and p-values associated with the top 50 DE genes. 

###Top 50 differentially expressed genes list
```{r topgenes}
##List of top 50 differentially expressed genes
topgenes <- topTags(et, n=50)
top.ann.genes <- ann.genes[row.names(topgenes),]
top50 <- cbind(topgenes[[1]], Symbol = top.ann.genes$symbol)
#knitr::kable(top50)

##Generating files required for top 50 DE gene plots
norm.count.sig50 <- normalized_counts[rownames(normalized_counts) %in% rownames(topgenes),]
ann.genes <- ann.genes[row.names(norm.count.sig50),]
row.names(norm.count.sig50) <- ann.genes[,8]
norm.count.sig50 <- as.matrix(norm.count.sig50)
ann.norm.count.sig50 <- as.data.frame(norm.count.sig50)
ann.norm.count.sig50 <- cbind(ann.norm.count.sig50, Symbol = ann.genes$symbol)
```
[Download the full top50 gene list](https://dl.dropboxusercontent.com/u/204381225/breault/results/top50_de_genes)

### Normalized counts plot for top 50 DE genes
The normalized counts plot visualizes the normalized counts for each replicate for each of the top 50 DE genes. This plot allows for visualization of the count variance between and within groups. The majority of the top 50 DE genes show clear differences in expression levels between the T24 and T7 groups. However, some of the samples exhibit very low counts for both groups. For example, gene Gm19345 appears as a single point on the graph, signifying that all replicates for both the T24 and T7 groups have very similar expression levels. Please note that many of these genes may be false positives due to the high FDR values (please see top50 gene list above).

```{r counts plot - top50 genes}
df <- data.frame(ann.norm.count.sig50)
df.m <- melt(df, id.vars = "Symbol")
ggplot(df.m, aes(x=Symbol, y=value, color = variable)) +
  geom_point() +
  scale_colour_manual(values = c("red", "red", "red", "blue", "blue", "blue", "blue"),
                      name = "Sample Groups",
                      breaks = c("S6", "S10"),
                      labels = c("Stressed lighting", "Control lighting")) +
        theme_bw() +
        theme(axis.text.x=element_text(angle = 45, hjust = 1)) +
        labs(title = "Normalized counts for top50 DE genes", x = "Differentially expressed genes", y = "Normalized counts per gene")
```

### Heatmap of top 50 DE genes
The heatmap for the top 50 DE genes show the sample groups to cluster well, with evident differences in expression between sample groups. 

```{r heatmap - top50 genes}
pheatmap(norm.count.sig50, annotation = meta, color = colorRampPalette(c("red", "orange", "light yellow"))(2000), cluster_rows = F, scale = "row")
```

# Functional Analysis for Top 50 DE Genes
Using the top 50 DE genes, a list of statistically enriched gene ontology (GO) terms was generated using the program [gprofileR](http://biit.cs.ut.ee/gprofiler/). The list of GO terms was then used as input to [REViGO](http://revigo.irb.hr/), which collapsed redundant and semantically-related terms, and output the most significantly enriched functional categories. The top 50 DE genes exhibit significant enrichment for processes related to metabolism and circadian rhythm. A [recent paper](https://dl.dropboxusercontent.com/u/204381225/breault/results/Science-2015-Zhang-1488-92.pdf) has explored the coupling of metabolism to circadian rhythm by NR1D1 (Rev-ErbA alpha).

```{r functional_analysis}
gene_list <- as.data.frame(top50)
gene_list <- cbind(gene_id = gene_list$Symbol, gene_list)
gene_list <- gene_list[, c(1,5)]

#gprofileR
gprofiler_results <- gprofiler(query = gene_list, organism = "mmusculus", ordered_query = F, 
    exclude_iea = F, max_set_size = 0, correction_method = "fdr", 
    hier_filtering = "none", domain_size = "annotated", custom_bg = "")

allterms <- gprofiler_results$term.id
GOs <- allterms[grep("GO:", allterms)]
pvals <- gprofiler_results$p.value[grep("GO:", allterms)]
GO.pval <- cbind(gprofiler_results$term.id, gprofiler_results$p.value)
GO_genes <- gprofiler_results[, c("term.id", "term.name", "p.value", "term.size", "overlap.size", "intersection")]
names(GO_genes) <- c("term.id", "term.name", "p.value", "term.size", "overlap.size", "assoc.gene.ids")
write.table(GO_genes, file ="GO_genes", quote = FALSE, sep = "\t", row.names=T)

runRevigo(GOs, pvals)
```
[Download GO Terms](https://dl.dropboxusercontent.com/u/204381225/breault/results/GO_terms_and_genes)

[Download REViGO Figure](https://dl.dropboxusercontent.com/u/204381225/breault/results/revigo_treemap.pdf)